// @assistant-ui/core/dist/store/runtime-clients/composer-runtime-client.js
import { resource, tapMemo, tapEffect, tapResource, withKey, } from "@assistant-ui/tap";
import { tapAssistantEmit, tapClientLookup, } from "@assistant-ui/store";
import { AttachmentRuntimeClient } from "./attachment-runtime-client.js";
import { tapSubscribable } from "./tap-subscribable.js";
const ComposerAttachmentClientByIndex = resource(({ runtime, index }) => {
    const attachmentRuntime = tapMemo(() => runtime.getAttachmentByIndex(index), [runtime, index]);
    return tapResource(AttachmentRuntimeClient({
        runtime: attachmentRuntime,
    }));
});
export const ComposerClient = resource(({ threadIdRef, messageIdRef, runtime, }) => {
    const runtimeState = tapSubscribable(runtime);
    const emit = tapAssistantEmit();
    // Bind composer events to event manager
    tapEffect(() => {
        const unsubscribers = [];
        // Subscribe to composer events
        for (const event of ["send", "attachmentAdd"]) {
            const unsubscribe = runtime.unstable_on(event, () => {
                emit(`composer.${event}`, {
                    threadId: threadIdRef.current,
                    ...(messageIdRef && { messageId: messageIdRef.current }),
                });
            });
            unsubscribers.push(unsubscribe);
        }
        // attachmentAddError carries the failed attachment ID
        unsubscribers.push(runtime.unstable_on("attachmentAddError", () => {
            const errorAttachment = runtime
                .getState()
                .attachments.findLast((a) => a.status.type === "incomplete" && a.status.reason === "error");
            emit("composer.attachmentAddError", {
                threadId: threadIdRef.current,
                ...(messageIdRef && { messageId: messageIdRef.current }),
                ...(errorAttachment && { attachmentId: errorAttachment.id }),
            });
        }));
        return () => {
            for (const unsub of unsubscribers)
                unsub();
        };
    }, [runtime, emit, threadIdRef, messageIdRef]);
    const attachments = tapClientLookup(() => runtimeState.attachments.map((attachment, idx) => withKey(attachment.id, ComposerAttachmentClientByIndex({
        runtime,
        index: idx,
    }))), [runtimeState.attachments, runtime]);
    const state = tapMemo(() => {
        return {
            text: runtimeState.text,
            role: runtimeState.role,
            attachments: attachments.state,
            runConfig: runtimeState.runConfig,
            isEditing: runtimeState.isEditing,
            canCancel: runtimeState.canCancel,
            attachmentAccept: runtimeState.attachmentAccept,
            isEmpty: runtimeState.isEmpty,
            type: runtimeState.type ?? "thread",
            dictation: runtimeState.dictation,
            quote: runtimeState.quote,
            queue: [],
        };
    }, [runtimeState, attachments.state]);
    return {
        getState: () => state,
        setText: runtime.setText,
        setRole: runtime.setRole,
        setRunConfig: runtime.setRunConfig,
        addAttachment: runtime.addAttachment,
        reset: runtime.reset,
        clearAttachments: runtime.clearAttachments,
        send: runtime.send,
        cancel: runtime.cancel,
        beginEdit: runtime.beginEdit ??
            (() => {
                throw new Error("beginEdit is not supported in this runtime");
            }),
        startDictation: runtime.startDictation,
        stopDictation: runtime.stopDictation,
        setQuote: runtime.setQuote,
        attachment: (selector) => {
            if ("id" in selector) {
                return attachments.get({ key: selector.id });
            }
            else {
                return attachments.get(selector);
            }
        },
        queueItem: () => {
            throw new Error("Queue is not supported in this runtime");
        },
        __internal_getRuntime: () => runtime,
    };
});
