// @assistant-ui/core/dist/store/runtime-clients/thread-list-item-runtime-client.js
import { resource, tapEffect } from "@assistant-ui/tap";
import { tapAssistantEmit } from "@assistant-ui/store";
import { tapSubscribable } from "./tap-subscribable.js";
export const ThreadListItemClient = resource(({ runtime, }) => {
    const state = tapSubscribable(runtime);
    const emit = tapAssistantEmit();
    // Bind thread list item events to event manager
    tapEffect(() => {
        const unsubscribers = [];
        // Subscribe to thread list item events
        const threadListItemEvents = [
            "switchedTo",
            "switchedAway",
        ];
        for (const event of threadListItemEvents) {
            const unsubscribe = runtime.unstable_on(event, () => {
                emit(`threadListItem.${event}`, {
                    threadId: runtime.getState().id,
                });
            });
            unsubscribers.push(unsubscribe);
        }
        return () => {
            for (const unsub of unsubscribers)
                unsub();
        };
    }, [runtime, emit]);
    return {
        getState: () => state,
        switchTo: runtime.switchTo,
        rename: runtime.rename,
        archive: runtime.archive,
        unarchive: runtime.unarchive,
        delete: runtime.delete,
        generateTitle: runtime.generateTitle,
        initialize: runtime.initialize,
        detach: runtime.detach,
        __internal_getRuntime: () => runtime,
    };
});
