// @assistant-ui/core/dist/model-context/frame/provider.js
import { toJSONSchema } from "assistant-stream";
import { FRAME_MESSAGE_CHANNEL, } from "./types.js";
const serializeTool = (tool) => ({
    ...(tool.description && { description: tool.description }),
    parameters: tool.parameters ? toJSONSchema(tool.parameters) : undefined,
    ...(tool.disabled !== undefined && { disabled: tool.disabled }),
    ...(tool.type && { type: tool.type }),
});
const serializeModelContext = (context) => ({
    ...(context.system !== undefined && { system: context.system }),
    ...(context.tools && {
        tools: Object.fromEntries(Object.entries(context.tools).map(([name, tool]) => [
            name,
            serializeTool(tool),
        ])),
    }),
});
export class AssistantFrameProvider {
    static _instance = null;
    _providers = new Set();
    _providerUnsubscribes = new Map();
    _targetOrigin;
    constructor(targetOrigin = "*") {
        this._targetOrigin = targetOrigin;
        this.handleMessage = this.handleMessage.bind(this);
        window.addEventListener("message", this.handleMessage);
        setTimeout(() => this.broadcastUpdate(), 0);
    }
    static getInstance(targetOrigin) {
        if (!AssistantFrameProvider._instance) {
            AssistantFrameProvider._instance = new AssistantFrameProvider(targetOrigin);
        }
        return AssistantFrameProvider._instance;
    }
    handleMessage(event) {
        if (this._targetOrigin !== "*" && event.origin !== this._targetOrigin)
            return;
        if (event.data?.channel !== FRAME_MESSAGE_CHANNEL)
            return;
        const message = event.data.message;
        switch (message.type) {
            case "model-context-request":
                this.sendMessage(event, {
                    type: "model-context-update",
                    context: serializeModelContext(this.getModelContext()),
                });
                break;
            case "tool-call":
                this.handleToolCall(message, event);
                break;
        }
    }
    async handleToolCall(message, event) {
        const tool = this.getModelContext().tools?.[message.toolName];
        let result;
        let error;
        if (!tool) {
            error = `Tool "${message.toolName}" not found`;
        }
        else {
            try {
                result = tool.execute
                    ? await tool.execute(message.args, {
                        toolCallId: message.id,
                        abortSignal: new AbortController().signal,
                        human: async () => {
                            throw new Error("Tool human input is not supported in frame context");
                        },
                    })
                    : undefined;
            }
            catch (e) {
                error = e instanceof Error ? e.message : String(e);
            }
        }
        this.sendMessage(event, {
            type: "tool-result",
            id: message.id,
            ...(error ? { error } : { result }),
        });
    }
    sendMessage(event, message) {
        event.source?.postMessage({ channel: FRAME_MESSAGE_CHANNEL, message }, { targetOrigin: event.origin });
    }
    getModelContext() {
        const contexts = Array.from(this._providers).map((p) => p.getModelContext());
        return contexts.reduce((merged, context) => ({
            system: context.system
                ? merged.system
                    ? `${merged.system}\n\n${context.system}`
                    : context.system
                : merged.system,
            tools: { ...(merged.tools || {}), ...(context.tools || {}) },
        }), {});
    }
    broadcastUpdate() {
        if (window.parent && window.parent !== window) {
            const updateMessage = {
                type: "model-context-update",
                context: serializeModelContext(this.getModelContext()),
            };
            window.parent.postMessage({ channel: FRAME_MESSAGE_CHANNEL, message: updateMessage }, this._targetOrigin);
        }
    }
    static addModelContextProvider(provider, targetOrigin) {
        const instance = AssistantFrameProvider.getInstance(targetOrigin);
        instance._providers.add(provider);
        const unsubscribe = provider.subscribe?.(() => instance.broadcastUpdate());
        if (unsubscribe) {
            instance._providerUnsubscribes.set(provider, unsubscribe);
        }
        instance.broadcastUpdate();
        return () => {
            instance._providers.delete(provider);
            instance._providerUnsubscribes.get(provider)?.();
            instance._providerUnsubscribes.delete(provider);
            instance.broadcastUpdate();
        };
    }
    static dispose() {
        if (AssistantFrameProvider._instance) {
            const instance = AssistantFrameProvider._instance;
            window.removeEventListener("message", instance.handleMessage);
            instance._providerUnsubscribes.forEach((unsubscribe) => unsubscribe?.());
            instance._providerUnsubscribes.clear();
            instance._providers.clear();
            AssistantFrameProvider._instance = null;
        }
    }
}
