// @unkey/api/dist/commonjs/funcs/keysRerollKey.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.keysRerollKey = keysRerollKey;
const encodings_js_1 = require("../lib/encodings.js");
const M = __importStar(require("../lib/matchers.js"));
const primitives_js_1 = require("../lib/primitives.js");
const schemas_js_1 = require("../lib/schemas.js");
const security_js_1 = require("../lib/security.js");
const url_js_1 = require("../lib/url.js");
const components = __importStar(require("../models/components/index.js"));
const errors = __importStar(require("../models/errors/index.js"));
const async_js_1 = require("../types/async.js");
/**
 * Reroll Key
 *
 * @remarks
 * Generate a new API key while preserving the configuration from an existing key.
 *
 * This operation creates a fresh key with a new token while maintaining all settings from the original key:
 * - Permissions and roles
 * - Custom metadata
 * - Rate limit configurations
 * - Identity associations
 * - Remaining credits
 * - Recovery settings
 *
 * **Key Generation:**
 * - The system attempts to extract the prefix from the original key
 * - If prefix extraction fails, the default API prefix is used
 * - Key length follows the API's default byte configuration (or 16 bytes if not specified)
 *
 * **Original Key Handling:**
 * - The original key will be revoked after the duration specified in `expiration`
 * - Set `expiration` to 0 to revoke immediately
 * - This allows for graceful key rotation with an overlap period
 *
 * Common use cases include:
 * - Rotating keys for security compliance
 * - Issuing replacement keys for compromised credentials
 * - Creating backup keys with identical permissions
 *
 * **Important:** Analytics and usage metrics are tracked at both the key level AND identity level. If the original key has an identity, the new key will inherit it, allowing you to track usage across both individual keys and the overall identity.
 *
 * **Required Permissions**
 *
 *  Your root key must have:
 *  - `api.*.create_key` or `api.<api_id>.create_key`
 *  - `api.*.encrypt_key` or `api.<api_id>.encrypt_key` (only when the original key is recoverable)
 */
function keysRerollKey(client, request, options) {
    return new async_js_1.APIPromise($do(client, request, options));
}
async function $do(client, request, options) {
    const parsed = (0, schemas_js_1.safeParse)(request, (value) => components.V2KeysRerollKeyRequestBody$outboundSchema.parse(value), "Input validation failed");
    if (!parsed.ok) {
        return [parsed, { status: "invalid" }];
    }
    const payload = parsed.value;
    const body = (0, encodings_js_1.encodeJSON)("body", payload, { explode: true });
    const path = (0, url_js_1.pathToFunc)("/v2/keys.rerollKey")();
    const headers = new Headers((0, primitives_js_1.compactMap)({
        "Content-Type": "application/json",
        Accept: "application/json",
    }));
    const secConfig = await (0, security_js_1.extractSecurity)(client._options.rootKey);
    const securityInput = secConfig == null ? {} : { rootKey: secConfig };
    const requestSecurity = (0, security_js_1.resolveGlobalSecurity)(securityInput);
    const context = {
        options: client._options,
        baseURL: options?.serverURL ?? client._baseURL ?? "",
        operationID: "keys.rerollKey",
        oAuth2Scopes: null,
        resolvedSecurity: requestSecurity,
        securitySource: client._options.rootKey,
        retryConfig: options?.retries
            || client._options.retryConfig
            || {
                strategy: "backoff",
                backoff: {
                    initialInterval: 50,
                    maxInterval: 1000,
                    exponent: 1.5,
                    maxElapsedTime: 10000,
                },
                retryConnectionErrors: true,
            }
            || { strategy: "none" },
        retryCodes: options?.retryCodes || ["5XX"],
    };
    const requestRes = client._createRequest(context, {
        security: requestSecurity,
        method: "POST",
        baseURL: options?.serverURL,
        path: path,
        headers: headers,
        body: body,
        userAgent: client._options.userAgent,
        timeoutMs: options?.timeoutMs || client._options.timeoutMs || -1,
    }, options);
    if (!requestRes.ok) {
        return [requestRes, { status: "invalid" }];
    }
    const req = requestRes.value;
    const doResult = await client._do(req, {
        context,
        errorCodes: ["400", "401", "403", "404", "429", "4XX", "500", "5XX"],
        retryConfig: context.retryConfig,
        retryCodes: context.retryCodes,
    });
    if (!doResult.ok) {
        return [doResult, { status: "request-error", request: req }];
    }
    const response = doResult.value;
    const responseFields = {
        HttpMeta: { Response: response, Request: req },
    };
    const [result] = await M.match(M.json(200, components.V2KeysRerollKeyResponseBody$inboundSchema), M.jsonErr(400, errors.BadRequestErrorResponse$inboundSchema), M.jsonErr(401, errors.UnauthorizedErrorResponse$inboundSchema), M.jsonErr(403, errors.ForbiddenErrorResponse$inboundSchema), M.jsonErr(404, errors.NotFoundErrorResponse$inboundSchema), M.jsonErr(429, errors.TooManyRequestsErrorResponse$inboundSchema, {
        ctype: "application/problem+json",
    }), M.jsonErr(500, errors.InternalServerErrorResponse$inboundSchema), M.fail("4XX"), M.fail("5XX"))(response, req, { extraFields: responseFields });
    if (!result.ok) {
        return [result, { status: "complete", request: req, response }];
    }
    return [result, { status: "complete", request: req, response }];
}
