// @assistant-ui/react/dist/unstable/useToolMentionAdapter.js
"use client";
import { useMemo } from "react";
import { useAui } from "@assistant-ui/store";
/**
 * @deprecated This API is still under active development and might change without notice.
 *
 * Creates a MentionAdapter for tools. When a user types `@`, they see
 * available tools and can mention them to hint the LLM to use a specific tool.
 *
 * @example
 * ```tsx
 * const mentionAdapter = unstable_useToolMentionAdapter({
 *   formatLabel: (name) => name.replaceAll("_", " "),
 * });
 * ```
 */
export function unstable_useToolMentionAdapter(options) {
    const aui = useAui();
    const explicitTools = options?.tools;
    const includeModelContext = options?.includeModelContextTools ?? !explicitTools;
    const formatLabel = options?.formatLabel;
    const categoryLabel = options?.categoryLabel;
    return useMemo(() => {
        const getTools = () => {
            const items = [];
            if (explicitTools) {
                items.push(...explicitTools);
            }
            if (includeModelContext) {
                const context = aui.thread().getModelContext();
                const tools = context.tools;
                if (tools) {
                    for (const [name, tool] of Object.entries(tools)) {
                        if (!items.some((i) => i.id === name)) {
                            items.push({
                                id: name,
                                type: "tool",
                                label: formatLabel ? formatLabel(name) : name,
                                description: tool.description ?? undefined,
                            });
                        }
                    }
                }
            }
            return items;
        };
        return {
            categories() {
                return [
                    { id: "tools", label: categoryLabel ?? "Tools", icon: undefined },
                ];
            },
            categoryItems(_categoryId) {
                return getTools();
            },
            search(query) {
                const lower = query.toLowerCase();
                return getTools().filter((item) => item.id.toLowerCase().includes(lower) ||
                    item.label.toLowerCase().includes(lower) ||
                    item.description?.toLowerCase().includes(lower));
            },
        };
    }, [aui, explicitTools, includeModelContext, formatLabel, categoryLabel]);
}
