// @unkey/api/dist/esm/models/errors/badrequesterrorresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod/v3";
import * as components from "../components/index.js";
import { UnkeyError } from "./unkeyerror.js";
/**
 * Error response for invalid requests that cannot be processed due to client-side errors. This typically occurs when request parameters are missing, malformed, or fail validation rules. The response includes detailed information about the specific errors in the request, including the location of each error and suggestions for fixing it. When receiving this error, check the 'errors' array in the response for specific validation issues that need to be addressed before retrying.
 */
export class BadRequestErrorResponse extends UnkeyError {
    constructor(err, httpMeta) {
        const message = "message" in err && typeof err.message === "string"
            ? err.message
            : `API error occurred: ${JSON.stringify(err)}`;
        super(message, httpMeta);
        this.data$ = err;
        this.meta = err.meta;
        this.error = err.error;
        this.name = "BadRequestErrorResponse";
    }
}
/** @internal */
export const BadRequestErrorResponse$inboundSchema = z.object({
    meta: components.Meta$inboundSchema,
    error: components.BadRequestErrorDetails$inboundSchema,
    request$: z.instanceof(Request),
    response$: z.instanceof(Response),
    body$: z.string(),
})
    .transform((v) => {
    return new BadRequestErrorResponse(v, {
        request: v.request$,
        response: v.response$,
        body: v.body$,
    });
});
