// @unkey/api/dist/esm/models/errors/goneerrorresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod/v3";
import * as components from "../components/index.js";
import { UnkeyError } from "./unkeyerror.js";
/**
 * Error response when the requested resource has been soft-deleted and is no longer available. This occurs when:
 *
 * @remarks
 * - The resource has been marked as deleted but still exists in the database
 * - The resource is intentionally unavailable but could potentially be restored
 * - The resource cannot be restored through the API or dashboard
 *
 * To resolve this error, contact support if you need the resource restored.
 */
export class GoneErrorResponse extends UnkeyError {
    constructor(err, httpMeta) {
        const message = "message" in err && typeof err.message === "string"
            ? err.message
            : `API error occurred: ${JSON.stringify(err)}`;
        super(message, httpMeta);
        this.data$ = err;
        this.meta = err.meta;
        this.error = err.error;
        this.name = "GoneErrorResponse";
    }
}
/** @internal */
export const GoneErrorResponse$inboundSchema = z.object({
    meta: components.Meta$inboundSchema,
    error: components.BaseError$inboundSchema,
    request$: z.instanceof(Request),
    response$: z.instanceof(Response),
    body$: z.string(),
})
    .transform((v) => {
    return new GoneErrorResponse(v, {
        request: v.request$,
        response: v.response$,
        body: v.body$,
    });
});
