// @unkey/api/dist/esm/models/errors/notfounderrorresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod/v3";
import * as components from "../components/index.js";
import { UnkeyError } from "./unkeyerror.js";
/**
 * Error response when the requested resource cannot be found. This occurs when:
 *
 * @remarks
 * - The specified resource ID doesn't exist in your workspace
 * - The resource has been deleted or moved
 * - The resource exists but is not accessible with current permissions
 *
 * To resolve this error, verify the resource ID is correct and that you have access to it.
 */
export class NotFoundErrorResponse extends UnkeyError {
    constructor(err, httpMeta) {
        const message = "message" in err && typeof err.message === "string"
            ? err.message
            : `API error occurred: ${JSON.stringify(err)}`;
        super(message, httpMeta);
        this.data$ = err;
        this.meta = err.meta;
        this.error = err.error;
        this.name = "NotFoundErrorResponse";
    }
}
/** @internal */
export const NotFoundErrorResponse$inboundSchema = z.object({
    meta: components.Meta$inboundSchema,
    error: components.BaseError$inboundSchema,
    request$: z.instanceof(Request),
    response$: z.instanceof(Response),
    body$: z.string(),
})
    .transform((v) => {
    return new NotFoundErrorResponse(v, {
        request: v.request$,
        response: v.response$,
        body: v.body$,
    });
});
