// @unkey/api/dist/esm/models/errors/serviceunavailableerrorresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod/v3";
import * as components from "../components/index.js";
import { UnkeyError } from "./unkeyerror.js";
/**
 * Error response when a required service is temporarily unavailable. This indicates that the service exists but cannot be reached or is not responding.
 *
 * @remarks
 *
 * When you encounter this error:
 * - The service is likely experiencing temporary issues
 * - Retrying the request after a short delay may succeed
 * - If the error persists, the service may be undergoing maintenance
 * - Contact Unkey support if the issue continues
 */
export class ServiceUnavailableErrorResponse extends UnkeyError {
    constructor(err, httpMeta) {
        const message = "message" in err && typeof err.message === "string"
            ? err.message
            : `API error occurred: ${JSON.stringify(err)}`;
        super(message, httpMeta);
        this.data$ = err;
        this.meta = err.meta;
        this.error = err.error;
        this.name = "ServiceUnavailableErrorResponse";
    }
}
/** @internal */
export const ServiceUnavailableErrorResponse$inboundSchema = z.object({
    meta: components.Meta$inboundSchema,
    error: components.BaseError$inboundSchema,
    request$: z.instanceof(Request),
    response$: z.instanceof(Response),
    body$: z.string(),
})
    .transform((v) => {
    return new ServiceUnavailableErrorResponse(v, {
        request: v.request$,
        response: v.response$,
        body: v.body$,
    });
});
