// @unkey/api/dist/esm/models/errors/unauthorizederrorresponse.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod/v3";
import * as components from "../components/index.js";
import { UnkeyError } from "./unkeyerror.js";
/**
 * Error response when authentication has failed or credentials are missing. This occurs when:
 *
 * @remarks
 * - No authentication token is provided in the request
 * - The provided token is invalid, expired, or malformed
 * - The token format doesn't match expected patterns
 *
 * To resolve this error, ensure you're including a valid root key in the Authorization header.
 */
export class UnauthorizedErrorResponse extends UnkeyError {
    constructor(err, httpMeta) {
        const message = "message" in err && typeof err.message === "string"
            ? err.message
            : `API error occurred: ${JSON.stringify(err)}`;
        super(message, httpMeta);
        this.data$ = err;
        this.meta = err.meta;
        this.error = err.error;
        this.name = "UnauthorizedErrorResponse";
    }
}
/** @internal */
export const UnauthorizedErrorResponse$inboundSchema = z.object({
    meta: components.Meta$inboundSchema,
    error: components.BaseError$inboundSchema,
    request$: z.instanceof(Request),
    response$: z.instanceof(Response),
    body$: z.string(),
})
    .transform((v) => {
    return new UnauthorizedErrorResponse(v, {
        request: v.request$,
        response: v.response$,
        body: v.body$,
    });
});
