// @unkey/api/dist/commonjs/models/errors/serviceunavailableerrorresponse.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceUnavailableErrorResponse$inboundSchema = exports.ServiceUnavailableErrorResponse = void 0;
const z = __importStar(require("zod/v3"));
const components = __importStar(require("../components/index.js"));
const unkeyerror_js_1 = require("./unkeyerror.js");
/**
 * Error response when a required service is temporarily unavailable. This indicates that the service exists but cannot be reached or is not responding.
 *
 * @remarks
 *
 * When you encounter this error:
 * - The service is likely experiencing temporary issues
 * - Retrying the request after a short delay may succeed
 * - If the error persists, the service may be undergoing maintenance
 * - Contact Unkey support if the issue continues
 */
class ServiceUnavailableErrorResponse extends unkeyerror_js_1.UnkeyError {
    constructor(err, httpMeta) {
        const message = "message" in err && typeof err.message === "string"
            ? err.message
            : `API error occurred: ${JSON.stringify(err)}`;
        super(message, httpMeta);
        this.data$ = err;
        this.meta = err.meta;
        this.error = err.error;
        this.name = "ServiceUnavailableErrorResponse";
    }
}
exports.ServiceUnavailableErrorResponse = ServiceUnavailableErrorResponse;
/** @internal */
exports.ServiceUnavailableErrorResponse$inboundSchema = z.object({
    meta: components.Meta$inboundSchema,
    error: components.BaseError$inboundSchema,
    request$: z.instanceof(Request),
    response$: z.instanceof(Response),
    body$: z.string(),
})
    .transform((v) => {
    return new ServiceUnavailableErrorResponse(v, {
        request: v.request$,
        response: v.response$,
        body: v.body$,
    });
});
