// @assistant-ui/react/dist/primitives/composer/mention/ComposerMentionContext.js
"use client";
import { jsx as _jsx } from "react/jsx-runtime";
import { createContext, useContext, useState, useCallback, useMemo, useEffect, } from "react";
import { useResource } from "@assistant-ui/tap/react";
import { useAui, useAuiState } from "@assistant-ui/store";
import { unstable_defaultDirectiveFormatter } from "@assistant-ui/core";
import { MentionResource, } from "./MentionResource.js";
// =============================================================================
// Context — public (popover components read state + actions from here)
// =============================================================================
const MentionContext = createContext(null);
export const useMentionContext = () => {
    const ctx = useContext(MentionContext);
    if (!ctx)
        throw new Error("useMentionContext must be used within ComposerPrimitive.MentionRoot");
    return ctx;
};
export const useMentionContextOptional = () => {
    return useContext(MentionContext);
};
const MentionInternalContext = createContext(null);
export const useMentionInternalContext = () => {
    return useContext(MentionInternalContext);
};
export const ComposerPrimitiveMentionRoot = ({ children, adapter: adapterProp, trigger: triggerChar = "@", formatter: formatterProp, }) => {
    const aui = useAui();
    const text = useAuiState((s) => s.composer.text);
    const formatter = formatterProp ?? unstable_defaultDirectiveFormatter;
    // ---------------------------------------------------------------------------
    // Runtime adapter (subscribe to state changes instead of useAuiState to avoid
    // infinite loop — getModelContext() returns a new object on every call)
    // ---------------------------------------------------------------------------
    const getRuntimeAdapter = useCallback(() => {
        try {
            const runtime = aui.composer().__internal_getRuntime?.();
            return runtime?._core?.getState()?.getMentionAdapter?.();
        }
        catch {
            return undefined;
        }
    }, [aui]);
    const [runtimeAdapter, setRuntimeAdapter] = useState(getRuntimeAdapter);
    useEffect(() => {
        return aui.subscribe(() => {
            setRuntimeAdapter((prev) => {
                const next = getRuntimeAdapter();
                return prev === next ? prev : next;
            });
        });
    }, [aui, getRuntimeAdapter]);
    const adapter = adapterProp ?? runtimeAdapter;
    // ---------------------------------------------------------------------------
    // Mention resource (all state + logic managed via tap primitives)
    // ---------------------------------------------------------------------------
    const mention = useResource(MentionResource({ adapter, text, triggerChar, formatter, aui }));
    // ---------------------------------------------------------------------------
    // Internal context (stable — methods come from tapEffectEvent)
    // ---------------------------------------------------------------------------
    const internalContextValue = useMemo(() => ({
        setCursorPosition: mention.setCursorPosition,
        registerSelectItemOverride: mention.registerSelectItemOverride,
    }), [mention.setCursorPosition, mention.registerSelectItemOverride]);
    return (_jsx(MentionContext.Provider, { value: mention, children: _jsx(MentionInternalContext.Provider, { value: internalContextValue, children: children }) }));
};
ComposerPrimitiveMentionRoot.displayName = "ComposerPrimitive.MentionRoot";
