// @assistant-ui/core/dist/index.js
// @assistant-ui/core - Framework-agnostic core runtime (public API)
export { mergeModelContexts } from "./model-context/types.js";
export { tool } from "./model-context/tool.js";
export { ModelContextRegistry } from "./model-context/registry.js";
export { AssistantFrameHost } from "./model-context/frame/host.js";
export { AssistantFrameProvider } from "./model-context/frame/provider.js";
export { FRAME_MESSAGE_CHANNEL } from "./model-context/frame/types.js";
export { SimpleImageAttachmentAdapter, SimpleTextAttachmentAdapter, CompositeAttachmentAdapter, } from "./adapters/attachment.js";
export { WebSpeechSynthesisAdapter, WebSpeechDictationAdapter, } from "./adapters/speech.js";
export { unstable_defaultDirectiveFormatter } from "./adapters/mention.js";
// External Store Message Utilities
export { getExternalStoreMessage, getExternalStoreMessages, bindExternalStoreMessage, } from "./runtime/utils/external-store-message.js";
export { ExportedMessageRepository } from "./runtime/utils/message-repository.js";
export { InMemoryThreadListAdapter } from "./runtimes/remote-thread-list/adapter/in-memory.js";
// Assistant Transport Utilities
export { toAISDKTools, getEnabledTools, createRequestHeaders, } from "./runtimes/assistant-transport/utils.js";
