// @assistant-ui/react/dist/client/ExternalThread.js
import { resource, tapState, tapMemo, tapEffect, tapEffectEvent, } from "@assistant-ui/tap";
import { tapClientLookup, attachTransformScopes, tapClientResource, Derived, } from "@assistant-ui/store";
import { withKey } from "@assistant-ui/tap";
import { ModelContext, Suggestions } from "@assistant-ui/core/store";
import { Tools, DataRenderers } from "@assistant-ui/core/react";
import { SingleThreadList } from "./SingleThreadList.js";
const EMPTY_QUEUE_ITEMS = [];
// Message Client - minimal implementation
const MessageClient = resource(({ message, index, onEdit, onReload, queue, }) => {
    const [isCopied, setIsCopied] = tapState(false);
    const [isHovering, setIsHovering] = tapState(false);
    const [isEditing, setIsEditing] = tapState(false);
    const partClients = tapClientLookup(() => message.content.map((part, idx) => withKey(idx, PartResource({ part }))), [message.content]);
    const attachmentClients = tapClientLookup(() => (message.attachments ?? []).map((attachment) => withKey(attachment.id, AttachmentResource({
        attachment,
        onRemove: () => { },
    }))), [message.attachments]);
    const handleBeginEdit = () => {
        setIsEditing(true);
    };
    const handleCancelEdit = () => {
        setIsEditing(false);
    };
    const handleSendEdit = (msg) => {
        queue?.clear("edit");
        onEdit?.({
            ...msg,
            parentId: message.id,
            sourceId: message.id,
        });
        setIsEditing(false);
    };
    const composerClient = tapClientResource(ComposerClientResource({
        type: "edit",
        isEditing,
        canCancel: true,
        onCancel: handleCancelEdit,
        onBeginEdit: handleBeginEdit,
        onSend: handleSendEdit,
        message,
        queue,
    }));
    const state = tapMemo(() => {
        return {
            ...message,
            attachments: message.attachments ?? [],
            parentId: null,
            isLast: false, // Will be set by thread
            branchNumber: 1,
            branchCount: 1,
            speech: undefined,
            submittedFeedback: undefined,
            parts: partClients.state,
            isCopied,
            isHovering,
            index,
            composer: composerClient.state,
        };
    }, [
        message,
        isCopied,
        isHovering,
        index,
        composerClient.state,
        partClients.state,
    ]);
    return {
        getState: () => state,
        composer: () => composerClient.methods,
        reload: () => {
            onReload?.();
        },
        speak: () => { },
        stopSpeaking: () => { },
        submitFeedback: () => { },
        switchToBranch: () => { },
        getCopyText: () => message.content.map((c) => ("text" in c ? c.text : "")).join(""),
        part: (selector) => {
            if ("index" in selector) {
                return partClients.get(selector);
            }
            const partIndex = state.parts.findIndex((p) => p.type === "tool-call" && p.toolCallId === selector.toolCallId);
            return partClients.get({ index: partIndex });
        },
        attachment: (selector) => {
            if ("id" in selector) {
                return attachmentClients.get({ key: selector.id });
            }
            return attachmentClients.get(selector);
        },
        setIsCopied,
        setIsHovering,
    };
});
// Part Client - minimal implementation
const PartResource = resource(({ part }) => {
    const state = tapMemo(() => ({
        ...part,
        status: { type: "complete" },
    }), [part]);
    return {
        getState: () => state,
        addToolResult: () => { },
        resumeToolCall: () => { },
    };
});
// Attachment Client - minimal implementation
const AttachmentResource = resource(({ attachment, onRemove, }) => {
    return {
        getState: () => attachment,
        remove: async () => {
            onRemove?.();
        },
    };
});
const QueueItemClient = resource(({ item, onSteer, onRemove, }) => {
    return {
        getState: () => item,
        steer: onSteer,
        remove: onRemove,
    };
});
// Composer Client - minimal implementation
const ComposerClientResource = resource(({ type, isEditing, canCancel, onCancel, onBeginEdit, onSend, message, queue, }) => {
    const [text, setText] = tapState("");
    const [role, setRole] = tapState("user");
    const [runConfig, setRunConfig] = tapState({});
    const [attachments, setAttachments] = tapState([]);
    const [quote, setQuote] = tapState(undefined);
    // Update composer values when editing begins
    const updateFromMessage = tapEffectEvent(() => {
        if (message) {
            // Extract text from message content (text parts only)
            const textParts = message.content.filter((part) => part.type === "text");
            const messageText = textParts
                .map((part) => ("text" in part ? part.text : ""))
                .join("\n\n");
            setText(messageText);
            setRole(message.role);
            setAttachments(message.attachments ?? []);
        }
    });
    tapEffect(() => {
        if (isEditing) {
            updateFromMessage();
        }
    }, [isEditing]);
    const attachmentClients = tapClientLookup(() => attachments.map((attachment, idx) => withKey(attachment.id, AttachmentResource({
        attachment,
        onRemove: () => {
            setAttachments(attachments.filter((_, i) => i !== idx));
        },
    }))), [attachments]);
    const queueItems = queue?.items ?? EMPTY_QUEUE_ITEMS;
    const queueItemClients = tapClientLookup(() => queueItems.map((item) => withKey(item.id, QueueItemClient({
        item,
        onSteer: () => queue?.steer(item.id),
        onRemove: () => queue?.remove(item.id),
    }))), [queueItems]);
    const state = tapMemo(() => ({
        text,
        role,
        attachments: attachmentClients.state,
        runConfig,
        isEditing,
        canCancel,
        attachmentAccept: "*",
        isEmpty: !text.trim() && !attachments.length,
        type,
        dictation: undefined,
        quote,
        queue: queueItems,
    }), [
        text,
        role,
        attachmentClients.state,
        runConfig,
        isEditing,
        canCancel,
        type,
        attachments.length,
        quote,
        queueItems,
    ]);
    return {
        getState: () => state,
        setText,
        setRole,
        setRunConfig,
        addAttachment: async (fileOrAttachment) => {
            if (fileOrAttachment instanceof File) {
                const newAttachment = {
                    id: Math.random().toString(36).substring(7),
                    type: "file",
                    name: fileOrAttachment.name,
                    contentType: fileOrAttachment.type,
                    file: fileOrAttachment,
                    status: { type: "complete" },
                    content: [],
                };
                setAttachments([...attachments, newAttachment]);
            }
            else {
                const newAttachment = {
                    id: fileOrAttachment.id ?? Math.random().toString(36).substring(7),
                    type: fileOrAttachment.type ?? "document",
                    name: fileOrAttachment.name,
                    contentType: fileOrAttachment.contentType,
                    content: fileOrAttachment.content,
                    status: { type: "complete" },
                };
                setAttachments([...attachments, newAttachment]);
            }
        },
        clearAttachments: async () => {
            setAttachments([]);
        },
        attachment: (selector) => {
            if ("id" in selector) {
                return attachmentClients.get({ key: selector.id });
            }
            return attachmentClients.get(selector);
        },
        reset: async () => {
            setText("");
            setRole("user");
            setRunConfig({});
            setAttachments([]);
            setQuote(undefined);
        },
        send: (opts) => {
            const currentQuote = quote;
            const composedMessage = {
                role,
                content: text ? [{ type: "text", text }] : [],
                attachments: attachments,
                createdAt: new Date(),
                parentId: null,
                sourceId: null,
                runConfig,
                metadata: {
                    custom: { ...(currentQuote ? { quote: currentQuote } : {}) },
                },
            };
            if (queue) {
                queue.enqueue(composedMessage, { steer: opts?.steer ?? false });
            }
            else {
                onSend?.(composedMessage);
            }
            setText("");
            setAttachments([]);
            setQuote(undefined);
        },
        cancel: onCancel,
        beginEdit: () => {
            onBeginEdit?.();
        },
        startDictation: () => { },
        stopDictation: () => { },
        setQuote,
        queueItem: (selector) => {
            return queueItemClients.get(selector);
        },
    };
});
// External Thread Client
export const ExternalThread = resource(({ messages, isRunning = false, onNew, onEdit, onReload, onStartRun, onCancel, queue, }) => {
    const handleReload = (messageId) => {
        const messageIndex = messages.findIndex((m) => m.id === messageId);
        if (messageIndex === -1)
            return;
        const parentId = messageIndex > 0 ? messages[messageIndex - 1].id : null;
        queue?.clear("reload");
        onReload?.(parentId);
    };
    const messageClients = tapClientLookup(() => messages.map((msg, index) => {
        const props = {
            message: msg,
            index,
            onReload: () => handleReload(msg.id),
            queue,
        };
        if (onEdit)
            props.onEdit = onEdit;
        return withKey(msg.id, MessageClient(props));
    }), [messages, onEdit, queue]);
    const handleCancelRun = () => {
        queue?.clear("cancel-run");
        onCancel?.();
    };
    const handleSendNew = (message) => {
        onNew?.(message);
    };
    const composerClient = tapClientResource(ComposerClientResource({
        type: "thread",
        isEditing: true,
        canCancel: isRunning,
        onCancel: handleCancelRun,
        onSend: handleSendNew,
        queue,
    }));
    const hasQueue = !!queue;
    const state = tapMemo(() => {
        const messageStates = messageClients.state.map((s, idx, arr) => ({
            ...s,
            isLast: idx === arr.length - 1,
        }));
        return {
            isEmpty: messages.length === 0,
            isDisabled: false,
            isLoading: false,
            isRunning,
            capabilities: {
                edit: false,
                reload: false,
                cancel: isRunning,
                speech: false,
                attachments: false,
                feedback: false,
                switchToBranch: false,
                switchBranchDuringRun: false,
                unstable_copy: false,
                dictation: false,
                queue: hasQueue,
            },
            messages: messageStates,
            state: {},
            suggestions: [],
            extras: undefined,
            speech: undefined,
            composer: composerClient.state,
        };
    }, [
        messages,
        isRunning,
        hasQueue,
        messageClients.state,
        composerClient.state,
    ]);
    return {
        getState: () => state,
        composer: () => composerClient.methods,
        append: (message) => {
            const appendMessage = typeof message === "string"
                ? {
                    createdAt: new Date(),
                    parentId: messages.at(-1)?.id ?? null,
                    sourceId: null,
                    runConfig: {},
                    role: "user",
                    content: [{ type: "text", text: message }],
                    attachments: [],
                    metadata: { custom: {} },
                }
                : {
                    createdAt: message.createdAt ?? new Date(),
                    parentId: message.parentId ?? messages.at(-1)?.id ?? null,
                    sourceId: message.sourceId ?? null,
                    role: message.role ?? "user",
                    content: message.content,
                    attachments: message.attachments ?? [],
                    metadata: message.metadata ?? { custom: {} },
                    runConfig: message.runConfig ?? {},
                    startRun: message.startRun,
                };
            if (queue) {
                queue.enqueue(appendMessage, { steer: false });
            }
            else {
                onNew?.(appendMessage);
            }
        },
        startRun: () => {
            onStartRun?.();
        },
        resumeRun: () => { },
        unstable_resumeRun: () => { },
        cancelRun: handleCancelRun,
        getModelContext: () => ({ tools: {}, config: {} }),
        export: () => ({ messages: [] }),
        import: () => { },
        reset: () => { },
        message: (selector) => {
            if ("id" in selector) {
                return messageClients.get({ key: selector.id });
            }
            return messageClients.get(selector);
        },
        stopSpeaking: () => { },
        startVoice: async () => { },
        stopVoice: async () => { },
    };
});
attachTransformScopes(ExternalThread, (scopes, parent) => {
    if (!scopes.threads && parent.threads.source === null) {
        const threadElement = scopes.thread;
        scopes.threads = SingleThreadList({ thread: threadElement });
        scopes.thread = Derived({
            source: "threads",
            query: { type: "main" },
            get: (aui) => aui.threads().thread("main"),
        });
    }
    if (!scopes.threadListItem && parent.threadListItem.source === null) {
        scopes.threadListItem = Derived({
            source: "threads",
            query: { type: "main" },
            get: (aui) => aui.threads().item("main"),
        });
    }
    scopes.composer ??= Derived({
        source: "thread",
        query: {},
        get: (aui) => aui.thread().composer(),
    });
    if (!scopes.modelContext && parent.modelContext.source === null) {
        scopes.modelContext = ModelContext();
    }
    if (!scopes.tools && parent.tools.source === null) {
        scopes.tools = Tools({});
    }
    if (!scopes.dataRenderers && parent.dataRenderers.source === null) {
        scopes.dataRenderers = DataRenderers();
    }
    if (!scopes.suggestions && parent.suggestions.source === null) {
        scopes.suggestions = Suggestions();
    }
});
