// @unkey/api/dist/esm/sdk/sdk.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import { ClientSDK } from "../lib/sdks.js";
import { Analytics } from "./analytics.js";
import { Apis } from "./apis.js";
import { Identities } from "./identities.js";
import { Internal } from "./internal.js";
import { Keys } from "./keys.js";
import { Permissions } from "./permissions.js";
import { Ratelimit } from "./ratelimit.js";
export class Unkey extends ClientSDK {
    get analytics() {
        return (this._analytics ?? (this._analytics = new Analytics(this._options)));
    }
    get apis() {
        return (this._apis ?? (this._apis = new Apis(this._options)));
    }
    get internal() {
        return (this._internal ?? (this._internal = new Internal(this._options)));
    }
    get identities() {
        return (this._identities ?? (this._identities = new Identities(this._options)));
    }
    get keys() {
        return (this._keys ?? (this._keys = new Keys(this._options)));
    }
    get permissions() {
        return (this._permissions ?? (this._permissions = new Permissions(this._options)));
    }
    get ratelimit() {
        return (this._ratelimit ?? (this._ratelimit = new Ratelimit(this._options)));
    }
}
