// @workos-inc/authkit-nextjs/dist/esm/middleware-helpers.js
import { NextResponse } from 'next/server';
/** Internal AuthKit headers - forwarded to downstream requests but never sent to browser. */
export const AUTHKIT_REQUEST_HEADERS = [
    'x-workos-middleware',
    'x-url',
    'x-redirect-uri',
    'x-sign-up-paths',
    'x-workos-session',
];
const ALLOWED_RESPONSE_HEADERS = [
    'set-cookie',
    'cache-control',
    'vary',
    'www-authenticate',
    'proxy-authenticate',
    'link',
    'x-middleware-cache',
];
const MULTI_VALUE_HEADERS = ['set-cookie', 'www-authenticate', 'proxy-authenticate', 'link'];
export function isAuthkitRequestHeader(name) {
    const lower = name.toLowerCase();
    return AUTHKIT_REQUEST_HEADERS.includes(lower) || lower.startsWith('x-workos-');
}
function setHeader(headers, name, value) {
    const lower = name.toLowerCase();
    if (MULTI_VALUE_HEADERS.includes(lower)) {
        headers.append(name, value);
    }
    else if (lower === 'vary') {
        const existing = headers.get(name);
        const merged = new Set([
            ...(existing ? existing.split(',').map((v) => v.trim()) : []),
            ...value.split(',').map((v) => v.trim()),
        ]);
        headers.set(name, [...merged].join(', '));
    }
    else {
        headers.set(name, value);
    }
}
/**
 * Partitions AuthKit headers into request headers (for withAuth) and response headers (for browser).
 */
export function partitionAuthkitHeaders(request, authkitHeaders) {
    const headers = new Headers(authkitHeaders);
    const requestHeaders = new Headers(request.headers);
    // Snapshot keys before iterating, since we delete headers during the loop
    const requestHeaderKeys = Array.from(requestHeaders.keys());
    for (const name of requestHeaderKeys) {
        if (isAuthkitRequestHeader(name)) {
            requestHeaders.delete(name);
        }
    }
    for (const headerName of AUTHKIT_REQUEST_HEADERS) {
        const value = headers.get(headerName);
        if (value != null) {
            requestHeaders.set(headerName, value);
        }
    }
    // Build response headers from allowlist only
    const responseHeaders = new Headers();
    for (const [name, value] of headers) {
        const lower = name.toLowerCase();
        if (!isAuthkitRequestHeader(lower) && ALLOWED_RESPONSE_HEADERS.includes(lower)) {
            setHeader(responseHeaders, name, value);
        }
    }
    // Auto-add cache-control when setting cookies
    if (responseHeaders.has('set-cookie') && !responseHeaders.has('cache-control')) {
        responseHeaders.set('cache-control', 'no-store');
    }
    return { requestHeaders, responseHeaders };
}
export function applyResponseHeaders(response, responseHeaders) {
    for (const [name, value] of responseHeaders) {
        setHeader(response.headers, name, value);
    }
    return response;
}
/**
 * Creates a NextResponse with properly merged AuthKit headers.
 */
export function handleAuthkitProxy(request, authkitHeaders, options = {}) {
    const { requestHeaders, responseHeaders } = partitionAuthkitHeaders(request, authkitHeaders);
    const { redirect, redirectStatus } = options;
    if (redirect != null && redirect !== '') {
        let redirectUrl;
        try {
            redirectUrl = redirect instanceof URL ? redirect : new URL(redirect, request.url);
        }
        catch {
            throw new Error(`Invalid redirect URL: "${redirect}". Must be a valid absolute or relative URL.`);
        }
        const method = request.method.toUpperCase();
        const status = redirectStatus ?? (method === 'GET' || method === 'HEAD' ? 307 : 303);
        return applyResponseHeaders(NextResponse.redirect(redirectUrl, status), responseHeaders);
    }
    return applyResponseHeaders(NextResponse.next({ request: { headers: requestHeaders } }), responseHeaders);
}
/** @deprecated Use `handleAuthkitProxy` instead. */
export const handleAuthkitHeaders = handleAuthkitProxy;
