// @workos-inc/authkit-nextjs/dist/esm/session.js
'use server';
import { sealData, unsealData } from 'iron-session';
import { createRemoteJWKSet, decodeJwt, jwtVerify } from 'jose';
import { cookies, headers } from 'next/headers';
import { redirect } from 'next/navigation';
import { getCookieOptions, getJwtCookie, getPKCECookieOptions } from './cookie.js';
import { WORKOS_CLIENT_ID, WORKOS_COOKIE_NAME, WORKOS_COOKIE_PASSWORD, WORKOS_REDIRECT_URI } from './env-variables.js';
import { TokenRefreshError, getSessionErrorContext } from './errors.js';
import { getAuthorizationUrl } from './get-authorization-url.js';
import { PKCE_COOKIE_NAME, setPKCECookie } from './pkce.js';
import { getWorkOS } from './workos.js';
import { parse, tokensToRegexp } from 'path-to-regexp';
import { handleAuthkitHeaders } from './middleware-helpers.js';
import { lazy, setCachePreventionHeaders } from './utils.js';
function appendPKCESetCookieHeader(headers, sealedState, requestUrl) {
    headers.append('Set-Cookie', `${PKCE_COOKIE_NAME}=${sealedState}; ${getPKCECookieOptions(requestUrl, true)}`);
}
const sessionHeaderName = 'x-workos-session';
const middlewareHeaderName = 'x-workos-middleware';
const signUpPathsHeaderName = 'x-sign-up-paths';
const jwtCookieName = 'workos-access-token';
const JWKS = lazy(() => createRemoteJWKSet(new URL(getWorkOS().userManagement.getJwksUrl(WORKOS_CLIENT_ID))));
/**
 * Applies cache security headers with Vary header deduplication.
 * Only applies headers if the request is authenticated (has session, cookie, or Authorization header).
 * Used in middleware where existing Vary headers may already be present.
 * @param headers - The Headers object to set the cache security headers on.
 * @param request - The NextRequest object to check for authentication.
 * @param sessionData - Optional session data to check for authentication.
 */
function applyCacheSecurityHeaders(headers, request, sessionData) {
    const cookieName = WORKOS_COOKIE_NAME || 'wos-session';
    // Only apply cache headers for authenticated requests
    if (!sessionData?.accessToken && !request.cookies.has(cookieName) && !request.headers.has('authorization')) {
        return;
    }
    const varyValues = new Set(['cookie']);
    if (request.headers.has('authorization')) {
        varyValues.add('authorization');
    }
    const currentVary = headers.get('Vary');
    if (currentVary) {
        currentVary.split(',').forEach((v) => {
            const trimmed = v.trim().toLowerCase();
            if (trimmed)
                varyValues.add(trimmed);
        });
    }
    headers.set('Vary', Array.from(varyValues)
        .map((v) => v.charAt(0).toUpperCase() + v.slice(1))
        .join(', '));
    setCachePreventionHeaders(headers);
}
/**
 * Determines if a request is for an initial document load (not API/RSC/prefetch)
 */
function isInitialDocumentRequest(request) {
    const accept = request.headers.get('accept') || '';
    const isDocumentRequest = accept.includes('text/html');
    const isRSCRequest = request.headers.has('RSC') || request.headers.has('Next-Router-State-Tree');
    const isPrefetch = request.headers.get('Purpose') === 'prefetch' ||
        request.headers.get('Sec-Purpose') === 'prefetch' ||
        request.headers.has('Next-Router-Prefetch');
    return isDocumentRequest && !isRSCRequest && !isPrefetch;
}
async function encryptSession(session) {
    return sealData(session, {
        password: WORKOS_COOKIE_PASSWORD,
        ttl: 0,
    });
}
async function updateSessionMiddleware(request, debug, middlewareAuth, redirectUri, signUpPaths, eagerAuth = false) {
    if (!redirectUri && !WORKOS_REDIRECT_URI) {
        throw new Error('You must provide a redirect URI in the AuthKit middleware or in the environment variables.');
    }
    if (!WORKOS_COOKIE_PASSWORD || WORKOS_COOKIE_PASSWORD.length < 32) {
        throw new Error('You must provide a valid cookie password that is at least 32 characters in the environment variables.');
    }
    let url;
    if (redirectUri) {
        url = new URL(redirectUri);
    }
    else {
        url = new URL(WORKOS_REDIRECT_URI);
    }
    if (middlewareAuth.enabled &&
        url.pathname === request.nextUrl.pathname &&
        !middlewareAuth.unauthenticatedPaths.includes(url.pathname)) {
        // In the case where:
        // - We're using middleware auth mode
        // - The redirect URI is in the middleware matcher
        // - The redirect URI isn't in the unauthenticatedPaths array
        //
        // then we would get stuck in a login loop due to the redirect happening before the session is set.
        // It's likely that the user accidentally forgot to add the path to unauthenticatedPaths, so we add it here.
        middlewareAuth.unauthenticatedPaths.push(url.pathname);
    }
    const matchedPaths = middlewareAuth.unauthenticatedPaths.filter((pathGlob) => {
        const pathRegex = getMiddlewareAuthPathRegex(pathGlob);
        return pathRegex.exec(request.nextUrl.pathname);
    });
    const { session, headers, authorizationUrl } = await updateSession(request, {
        debug,
        redirectUri,
        screenHint: getScreenHint(signUpPaths, request.nextUrl.pathname),
        eagerAuth,
    });
    // Record the sign up paths so we can use them later
    if (signUpPaths.length > 0) {
        headers.set(signUpPathsHeaderName, signUpPaths.join(','));
    }
    applyCacheSecurityHeaders(headers, request, session);
    // If the user is logged out and this path isn't on the allowlist for logged out paths, redirect to AuthKit.
    if (middlewareAuth.enabled && matchedPaths.length === 0 && !session.user) {
        if (debug) {
            console.log(`Unauthenticated user on protected route ${request.url}, redirecting to AuthKit`);
        }
        return handleAuthkitHeaders(request, headers, { redirect: authorizationUrl });
    }
    return handleAuthkitHeaders(request, headers);
}
async function updateSession(request, options = { debug: false }) {
    const session = await getSessionFromCookie(request);
    // Since we're setting the headers in the response, we need to create a new Headers object without copying
    // the request headers.
    // See https://github.com/vercel/next.js/issues/50659#issuecomment-2333990159
    const newRequestHeaders = new Headers();
    // Record that the request was routed through the middleware so we can check later for DX purposes
    newRequestHeaders.set(middlewareHeaderName, 'true');
    // We store the current request url in a custom header, so we can always have access to it
    // This is because on hard navigations we don't have access to `next-url` but need to get the current
    // `pathname` to be able to return the users where they came from before sign-in
    newRequestHeaders.set('x-url', request.url);
    if (options.redirectUri) {
        // Store the redirect URI in a custom header, so we always have access to it and so that subsequent
        // calls to `getAuthorizationUrl` will use the same redirect URI
        newRequestHeaders.set('x-redirect-uri', options.redirectUri);
    }
    newRequestHeaders.delete(sessionHeaderName);
    if (!session) {
        if (options.debug) {
            console.log('No session found from cookie');
        }
        const { url: authorizationUrl, sealedState } = await getAuthorizationUrl({
            returnPathname: getReturnPathname(request.url),
            redirectUri: options.redirectUri || WORKOS_REDIRECT_URI,
            screenHint: options.screenHint,
        });
        appendPKCESetCookieHeader(newRequestHeaders, sealedState, request.url);
        return {
            session: { user: null },
            headers: newRequestHeaders,
            authorizationUrl,
        };
    }
    const hasValidSession = await verifyAccessToken(session.accessToken);
    const cookieName = WORKOS_COOKIE_NAME || 'wos-session';
    applyCacheSecurityHeaders(newRequestHeaders, request, session);
    if (hasValidSession) {
        newRequestHeaders.set(sessionHeaderName, request.cookies.get(cookieName).value);
        const { sid: sessionId, org_id: organizationId, role, roles, permissions, entitlements, feature_flags: featureFlags, } = decodeJwt(session.accessToken);
        // Set JWT cookie if eagerAuth is enabled
        // Only set on document requests (initial page loads), not on API/RSC requests
        if (options.eagerAuth && isInitialDocumentRequest(request)) {
            const existingJwtCookie = request.cookies.get(jwtCookieName);
            // Only set if cookie doesn't exist or has different value
            if (!existingJwtCookie || existingJwtCookie.value !== session.accessToken) {
                newRequestHeaders.append('Set-Cookie', getJwtCookie(session.accessToken, request.url));
            }
        }
        return {
            session: {
                sessionId,
                user: session.user,
                organizationId,
                role,
                roles,
                permissions,
                entitlements,
                featureFlags,
                impersonator: session.impersonator,
                accessToken: session.accessToken,
            },
            headers: newRequestHeaders,
        };
    }
    try {
        if (options.debug) {
            // istanbul ignore next
            console.log(`Session invalid. ${session.accessToken ? `Refreshing access token that ends in ${session.accessToken.slice(-10)}` : 'Access token missing.'}`);
        }
        const { org_id: organizationIdFromAccessToken } = decodeJwt(session.accessToken);
        const { accessToken, refreshToken, user, impersonator } = await getWorkOS().userManagement.authenticateWithRefreshToken({
            clientId: WORKOS_CLIENT_ID,
            refreshToken: session.refreshToken,
            organizationId: organizationIdFromAccessToken,
        });
        if (options.debug) {
            console.log('Session successfully refreshed');
        }
        // Encrypt session with new access and refresh tokens
        const encryptedSession = await encryptSession({
            accessToken,
            refreshToken,
            user,
            impersonator,
        });
        newRequestHeaders.append('Set-Cookie', `${cookieName}=${encryptedSession}; ${getCookieOptions(request.url, true)}`);
        newRequestHeaders.set(sessionHeaderName, encryptedSession);
        // Set JWT cookie if eagerAuth is enabled
        // Only set on document requests (initial page loads), not on API/RSC requests
        if (options.eagerAuth && isInitialDocumentRequest(request)) {
            newRequestHeaders.append('Set-Cookie', getJwtCookie(accessToken, request.url));
        }
        const { sid: sessionId, org_id: organizationId, role, roles, permissions, entitlements, feature_flags: featureFlags, } = decodeJwt(accessToken);
        options.onSessionRefreshSuccess?.({ accessToken, user, impersonator, organizationId });
        return {
            session: {
                sessionId,
                user,
                organizationId,
                role,
                roles,
                permissions,
                entitlements,
                featureFlags,
                impersonator,
                accessToken,
            },
            headers: newRequestHeaders,
        };
    }
    catch (e) {
        if (options.debug) {
            console.log('Failed to refresh. Deleting cookie.', e);
        }
        // When we need to delete a cookie, return it as a header as you can't delete cookies from edge middleware
        const deleteCookie = `${cookieName}=; Expires=${new Date(0).toUTCString()}; ${getCookieOptions(request.url, true, true)}`;
        newRequestHeaders.append('Set-Cookie', deleteCookie);
        // Delete JWT cookie if eagerAuth is enabled
        if (options.eagerAuth) {
            const deleteJwtCookie = getJwtCookie(null, request.url, true);
            newRequestHeaders.append('Set-Cookie', deleteJwtCookie);
        }
        options.onSessionRefreshError?.({ error: e, request });
        const { url: authorizationUrl, sealedState } = await getAuthorizationUrl({
            returnPathname: getReturnPathname(request.url),
            redirectUri: options.redirectUri || WORKOS_REDIRECT_URI,
        });
        appendPKCESetCookieHeader(newRequestHeaders, sealedState, request.url);
        return {
            session: { user: null },
            headers: newRequestHeaders,
            authorizationUrl,
        };
    }
}
async function refreshSession({ organizationId: nextOrganizationId, ensureSignedIn = false, } = {}) {
    const session = await getSessionFromCookie();
    if (!session) {
        if (ensureSignedIn) {
            await redirectToSignIn();
        }
        return { user: null };
    }
    const { org_id: organizationIdFromAccessToken } = decodeJwt(session.accessToken);
    let refreshResult;
    try {
        refreshResult = await getWorkOS().userManagement.authenticateWithRefreshToken({
            clientId: WORKOS_CLIENT_ID,
            refreshToken: session.refreshToken,
            organizationId: nextOrganizationId ?? organizationIdFromAccessToken,
        });
    }
    catch (error) {
        throw new TokenRefreshError(`Failed to refresh session: ${error instanceof Error ? error.message : String(error)}`, error, getSessionErrorContext(session));
    }
    const headersList = await headers();
    const url = headersList.get('x-url');
    await saveSession(refreshResult, url || WORKOS_REDIRECT_URI);
    const { accessToken, user, impersonator } = refreshResult;
    const { sid: sessionId, org_id: organizationId, role, roles, permissions, entitlements, feature_flags: featureFlags, } = decodeJwt(accessToken);
    return {
        sessionId,
        user,
        organizationId,
        role,
        roles,
        permissions,
        entitlements,
        featureFlags,
        impersonator,
        accessToken,
    };
}
function getMiddlewareAuthPathRegex(pathGlob) {
    try {
        const url = new URL(pathGlob, 'https://example.com');
        const path = `${url.pathname}${url.hash || ''}`;
        const tokens = parse(path);
        const regex = tokensToRegexp(tokens).source;
        return new RegExp(regex);
    }
    catch (err) {
        console.log('err', err);
        const message = err instanceof Error ? err.message : String(err);
        throw new Error(`Error parsing routes for middleware auth. Reason: ${message}`);
    }
}
async function redirectToSignIn() {
    const headersList = await headers();
    const url = headersList.get('x-url');
    if (!url) {
        throw new Error('No URL found in the headers');
    }
    // Determine if the current route is in the sign up paths
    const signUpPaths = headersList.get(signUpPathsHeaderName)?.split(',');
    const pathname = new URL(url).pathname;
    const screenHint = getScreenHint(signUpPaths, pathname);
    const returnPathname = getReturnPathname(url);
    const { url: authkitUrl, sealedState } = await getAuthorizationUrl({ returnPathname, screenHint });
    await setPKCECookie(sealedState);
    redirect(authkitUrl);
}
export async function getTokenClaims(accessToken) {
    const token = accessToken ?? (await withAuth()).accessToken;
    if (!token) {
        return {};
    }
    return decodeJwt(token);
}
async function withAuth(options) {
    const session = await getSessionFromHeader();
    if (!session) {
        if (options?.ensureSignedIn) {
            await redirectToSignIn();
        }
        return { user: null };
    }
    const { sid: sessionId, org_id: organizationId, role, roles, permissions, entitlements, feature_flags: featureFlags, } = decodeJwt(session.accessToken);
    return {
        sessionId,
        user: session.user,
        organizationId,
        role,
        roles,
        permissions,
        entitlements,
        featureFlags,
        impersonator: session.impersonator,
        accessToken: session.accessToken,
    };
}
async function verifyAccessToken(accessToken) {
    try {
        await jwtVerify(accessToken, JWKS());
        return true;
    }
    catch {
        return false;
    }
}
export async function getSessionFromCookie(request) {
    const cookieName = WORKOS_COOKIE_NAME || 'wos-session';
    let cookie;
    if (request) {
        cookie = request.cookies.get(cookieName);
    }
    else {
        const nextCookies = await cookies();
        cookie = nextCookies.get(cookieName);
    }
    if (cookie) {
        return unsealData(cookie.value, {
            password: WORKOS_COOKIE_PASSWORD,
        });
    }
}
async function getSessionFromHeader() {
    const headersList = await headers();
    const hasMiddleware = Boolean(headersList.get(middlewareHeaderName));
    if (!hasMiddleware) {
        const url = headersList.get('x-url');
        throw new Error(`You are calling 'withAuth' on ${url ?? 'a route'} that isn't covered by the AuthKit middleware. Make sure it is running on all paths you are calling 'withAuth' from by updating your middleware config in 'middleware.(js|ts)'.`);
    }
    const authHeader = headersList.get(sessionHeaderName);
    if (!authHeader)
        return;
    return unsealData(authHeader, { password: WORKOS_COOKIE_PASSWORD });
}
function getReturnPathname(url) {
    const newUrl = new URL(url);
    return `${newUrl.pathname}${newUrl.search}`;
}
function getScreenHint(signUpPaths, pathname) {
    if (!signUpPaths)
        return 'sign-in';
    const screenHintPaths = signUpPaths.filter((pathGlob) => {
        const pathRegex = getMiddlewareAuthPathRegex(pathGlob);
        return pathRegex.exec(pathname);
    });
    return screenHintPaths.length > 0 ? 'sign-up' : 'sign-in';
}
/**
 * Saves a WorkOS session to a cookie for use with AuthKit.
 *
 * This function is intended for advanced use cases where you need to manually manage sessions,
 * such as custom authentication flows (email verification, etc.) that don't use
 * the standard AuthKit authentication flow.
 *
 * @param sessionOrResponse The WorkOS session or AuthenticationResponse containing access token, refresh token, and user information.
 * @param request Either a NextRequest object or a URL string, used to determine cookie settings.
 *
 * @example
 * // With a NextRequest object
 * import { saveSession } from '@workos-inc/authkit-nextjs';
 *
 * async function handleEmailVerification(req: NextRequest) {
 *   const { code } = await req.json();
 *   const authResponse = await workos.userManagement.authenticateWithEmailVerification({
 *     clientId: process.env.WORKOS_CLIENT_ID,
 *     code,
 *   });
 *
 *   await saveSession(authResponse, req);
 * }
 *
 * @example
 * // With a URL string
 * await saveSession(authResponse, 'https://example.com/callback');
 */
export async function saveSession(sessionOrResponse, request) {
    const cookieName = WORKOS_COOKIE_NAME || 'wos-session';
    const encryptedSession = await encryptSession(sessionOrResponse);
    const nextCookies = await cookies();
    const url = typeof request === 'string' ? request : request.url;
    nextCookies.set(cookieName, encryptedSession, getCookieOptions(url));
}
export { encryptSession, refreshSession, updateSession, updateSessionMiddleware, withAuth };
