// @assistant-ui/core/dist/runtime/base/base-composer-runtime-core.js
import { BaseSubscribable } from "../../subscribable/subscribable.js";
import { fileMatchesAccept, } from "../../adapters/attachment.js";
import { generateId } from "../../utils/id.js";
const isAttachmentComplete = (a) => a.status.type === "complete";
export class BaseComposerRuntimeCore extends BaseSubscribable {
    isEditing = true;
    get attachmentAccept() {
        return this.getAttachmentAdapter()?.accept ?? "*";
    }
    _attachments = [];
    get attachments() {
        return this._attachments;
    }
    setAttachments(value) {
        this._attachments = value;
        this._notifySubscribers();
    }
    get isEmpty() {
        return !this.text.trim() && !this.attachments.length;
    }
    _text = "";
    get text() {
        return this._text;
    }
    _role = "user";
    get role() {
        return this._role;
    }
    _runConfig = {};
    get runConfig() {
        return this._runConfig;
    }
    _quote = undefined;
    get quote() {
        return this._quote;
    }
    setQuote(quote) {
        if (this._quote === quote)
            return;
        this._quote = quote;
        this._notifySubscribers();
    }
    setText(value) {
        if (this._text === value)
            return;
        this._text = value;
        if (this._dictation) {
            this._dictationBaseText = value;
            this._currentInterimText = "";
            const { status, inputDisabled } = this._dictation;
            this._dictation = inputDisabled ? { status, inputDisabled } : { status };
        }
        this._notifySubscribers();
    }
    setRole(role) {
        if (this._role === role)
            return;
        this._role = role;
        this._notifySubscribers();
    }
    setRunConfig(runConfig) {
        if (this._runConfig === runConfig)
            return;
        this._runConfig = runConfig;
        this._notifySubscribers();
    }
    _emptyTextAndAttachments() {
        this._attachments = [];
        this._text = "";
        this._notifySubscribers();
    }
    async _onClearAttachments() {
        const adapter = this.getAttachmentAdapter();
        if (adapter) {
            const pending = this._attachments.filter((a) => !isAttachmentComplete(a));
            await Promise.all(pending.map((a) => adapter.remove(a)));
        }
    }
    async reset() {
        if (this._attachments.length === 0 &&
            this._text === "" &&
            this._role === "user" &&
            Object.keys(this._runConfig).length === 0 &&
            this._quote === undefined) {
            return;
        }
        this._role = "user";
        this._runConfig = {};
        this._quote = undefined;
        const task = this._onClearAttachments();
        this._emptyTextAndAttachments();
        await task;
    }
    async clearAttachments() {
        const task = this._onClearAttachments();
        this.setAttachments([]);
        await task;
    }
    async send() {
        if (this.isEmpty)
            return;
        if (this._dictationSession) {
            this._dictationSession.cancel();
            this._cleanupDictation();
        }
        const adapter = this.getAttachmentAdapter();
        const attachments = this.attachments.length > 0
            ? Promise.all(this.attachments.map(async (a) => {
                if (isAttachmentComplete(a))
                    return a;
                if (!adapter)
                    throw new Error("Attachments are not supported");
                const result = await adapter.send(a);
                return result;
            }))
            : [];
        const text = this.text;
        const quote = this._quote;
        this._quote = undefined;
        this._emptyTextAndAttachments();
        const message = {
            createdAt: new Date(),
            role: this.role,
            content: text ? [{ type: "text", text }] : [],
            attachments: await attachments,
            runConfig: this.runConfig,
            metadata: { custom: { ...(quote ? { quote } : {}) } },
        };
        this.handleSend(message);
        this._notifyEventSubscribers("send");
    }
    cancel() {
        this.handleCancel();
    }
    async addAttachment(fileOrAttachment) {
        if (!(fileOrAttachment instanceof File)) {
            const a = {
                id: fileOrAttachment.id ?? generateId(),
                type: fileOrAttachment.type ?? "document",
                name: fileOrAttachment.name,
                contentType: fileOrAttachment.contentType,
                content: fileOrAttachment.content,
                status: { type: "complete" },
            };
            this._attachments = [...this._attachments, a];
            this._notifyEventSubscribers("attachmentAdd");
            this._notifySubscribers();
            return;
        }
        const adapter = this.getAttachmentAdapter();
        if (!adapter)
            throw new Error("Attachments are not supported");
        if (!fileMatchesAccept({ name: fileOrAttachment.name, type: fileOrAttachment.type }, adapter.accept)) {
            throw new Error(`File type ${fileOrAttachment.type || "unknown"} is not accepted. Accepted types: ${adapter.accept}`);
        }
        const upsertAttachment = (a) => {
            const idx = this._attachments.findIndex((attachment) => attachment.id === a.id);
            if (idx !== -1)
                this._attachments = [
                    ...this._attachments.slice(0, idx),
                    a,
                    ...this._attachments.slice(idx + 1),
                ];
            else {
                this._attachments = [...this._attachments, a];
            }
            this._notifySubscribers();
        };
        let lastAttachment;
        try {
            const promiseOrGenerator = adapter.add({ file: fileOrAttachment });
            if (Symbol.asyncIterator in promiseOrGenerator) {
                for await (const r of promiseOrGenerator) {
                    lastAttachment = r;
                    upsertAttachment(r);
                }
            }
            else {
                lastAttachment = await promiseOrGenerator;
                upsertAttachment(lastAttachment);
            }
        }
        catch (e) {
            if (lastAttachment) {
                upsertAttachment({
                    ...lastAttachment,
                    status: { type: "incomplete", reason: "error" },
                });
            }
            try {
                this._notifyEventSubscribers("attachmentAddError");
            }
            catch {
                // prevent subscriber errors from masking the adapter error
            }
            throw e;
        }
        const hasError = lastAttachment?.status.type === "incomplete" &&
            lastAttachment.status.reason === "error";
        this._notifyEventSubscribers(hasError ? "attachmentAddError" : "attachmentAdd");
    }
    async removeAttachment(attachmentId) {
        const index = this._attachments.findIndex((a) => a.id === attachmentId);
        if (index === -1)
            throw new Error("Attachment not found");
        const attachment = this._attachments[index];
        if (!isAttachmentComplete(attachment)) {
            const adapter = this.getAttachmentAdapter();
            if (!adapter)
                throw new Error("Attachments are not supported");
            await adapter.remove(attachment);
        }
        this._attachments = this._attachments.filter((a) => a.id !== attachmentId);
        this._notifySubscribers();
    }
    _dictation;
    _dictationSession;
    _dictationUnsubscribes = [];
    _dictationBaseText = "";
    _currentInterimText = "";
    _dictationSessionIdCounter = 0;
    _activeDictationSessionId;
    _isCleaningDictation = false;
    get dictation() {
        return this._dictation;
    }
    _isActiveSession(sessionId, session) {
        return (this._activeDictationSessionId === sessionId &&
            this._dictationSession === session);
    }
    startDictation() {
        const adapter = this.getDictationAdapter();
        if (!adapter) {
            throw new Error("Dictation adapter not configured");
        }
        if (this._dictationSession) {
            for (const unsub of this._dictationUnsubscribes) {
                unsub();
            }
            this._dictationUnsubscribes = [];
            const oldSession = this._dictationSession;
            oldSession.stop().catch(() => { });
            this._dictationSession = undefined;
        }
        const inputDisabled = adapter.disableInputDuringDictation ?? false;
        this._dictationBaseText = this._text;
        this._currentInterimText = "";
        const session = adapter.listen();
        this._dictationSession = session;
        const sessionId = ++this._dictationSessionIdCounter;
        this._activeDictationSessionId = sessionId;
        this._dictation = { status: session.status, inputDisabled };
        this._notifySubscribers();
        const unsubSpeech = session.onSpeech((result) => {
            if (!this._isActiveSession(sessionId, session))
                return;
            const isFinal = result.isFinal !== false;
            const needsSeparator = this._dictationBaseText &&
                !this._dictationBaseText.endsWith(" ") &&
                result.transcript;
            const separator = needsSeparator ? " " : "";
            if (isFinal) {
                this._dictationBaseText =
                    this._dictationBaseText + separator + result.transcript;
                this._currentInterimText = "";
                this._text = this._dictationBaseText;
                if (this._dictation) {
                    const { transcript: _, ...rest } = this._dictation;
                    this._dictation = rest;
                }
                this._notifySubscribers();
            }
            else {
                this._currentInterimText = separator + result.transcript;
                this._text = this._dictationBaseText + this._currentInterimText;
                if (this._dictation) {
                    this._dictation = {
                        ...this._dictation,
                        transcript: result.transcript,
                    };
                }
                this._notifySubscribers();
            }
        });
        this._dictationUnsubscribes.push(unsubSpeech);
        const unsubStart = session.onSpeechStart(() => {
            if (!this._isActiveSession(sessionId, session))
                return;
            this._dictation = {
                status: { type: "running" },
                inputDisabled,
                ...(this._dictation?.transcript && {
                    transcript: this._dictation.transcript,
                }),
            };
            this._notifySubscribers();
        });
        this._dictationUnsubscribes.push(unsubStart);
        const unsubEnd = session.onSpeechEnd(() => {
            this._cleanupDictation({ sessionId });
        });
        this._dictationUnsubscribes.push(unsubEnd);
        const statusInterval = setInterval(() => {
            if (!this._isActiveSession(sessionId, session))
                return;
            if (session.status.type === "ended") {
                this._cleanupDictation({ sessionId });
            }
        }, 100);
        this._dictationUnsubscribes.push(() => clearInterval(statusInterval));
    }
    stopDictation() {
        if (!this._dictationSession)
            return;
        const session = this._dictationSession;
        const sessionId = this._activeDictationSessionId;
        session.stop().finally(() => {
            this._cleanupDictation({ sessionId });
        });
    }
    _cleanupDictation(options) {
        const isStaleSession = options?.sessionId !== undefined &&
            options.sessionId !== this._activeDictationSessionId;
        if (isStaleSession || this._isCleaningDictation)
            return;
        this._isCleaningDictation = true;
        try {
            for (const unsub of this._dictationUnsubscribes) {
                unsub();
            }
            this._dictationUnsubscribes = [];
            this._dictationSession = undefined;
            this._activeDictationSessionId = undefined;
            this._dictation = undefined;
            this._dictationBaseText = "";
            this._currentInterimText = "";
            this._notifySubscribers();
        }
        finally {
            this._isCleaningDictation = false;
        }
    }
    _eventSubscribers = new Map();
    _notifyEventSubscribers(event) {
        const subscribers = this._eventSubscribers.get(event);
        if (!subscribers)
            return;
        for (const callback of subscribers)
            callback();
    }
    unstable_on(event, callback) {
        const subscribers = this._eventSubscribers.get(event);
        if (!subscribers) {
            this._eventSubscribers.set(event, new Set([callback]));
        }
        else {
            subscribers.add(callback);
        }
        return () => {
            const subscribers = this._eventSubscribers.get(event);
            if (!subscribers)
                return;
            subscribers.delete(callback);
        };
    }
}
