// @assistant-ui/react/dist/index.js
/// <reference types="@assistant-ui/core/react" />
// Re-export from @assistant-ui/store
export { useAui, AuiProvider, useAuiState, useAuiEvent, AuiIf, } from "@assistant-ui/store";
// ============================================================================
// legacy-runtime/cloud
// ============================================================================
export { useCloudThreadListRuntime } from "./legacy-runtime/cloud/useCloudThreadListRuntime.js";
export { AssistantCloud } from "assistant-cloud";
export { SimpleImageAttachmentAdapter, SimpleTextAttachmentAdapter, CompositeAttachmentAdapter, } from "@assistant-ui/core";
export { CloudFileAttachmentAdapter } from "./legacy-runtime/runtime-cores/adapters/attachment/CloudFileAttachmentAdapter.js";
export { WebSpeechSynthesisAdapter, WebSpeechDictationAdapter, } from "@assistant-ui/core";
// --- adapters/RuntimeAdapterProvider ---
export { RuntimeAdapterProvider, useRuntimeAdapters, } from "@assistant-ui/core/react";
// --- assistant-transport ---
export { useAssistantTransportRuntime, useAssistantTransportSendCommand, useAssistantTransportState, } from "./legacy-runtime/runtime-cores/assistant-transport/useAssistantTransportRuntime.js";
export { getExternalStoreMessage, getExternalStoreMessages, bindExternalStoreMessage, } from "@assistant-ui/core";
export { useExternalStoreRuntime } from "./legacy-runtime/runtime-cores/external-store/useExternalStoreRuntime.js";
export { useExternalMessageConverter, convertExternalMessages as unstable_convertExternalMessages, } from "./legacy-runtime/runtime-cores/external-store/external-message-converter.js";
export { createMessageConverter as unstable_createMessageConverter } from "./legacy-runtime/runtime-cores/external-store/createMessageConverter.js";
export { useLocalRuntime } from "./legacy-runtime/runtime-cores/local/useLocalRuntime.js";
/**
 * @deprecated Use `useLocalRuntime` instead.
 */
export { useLocalRuntime as useLocalThreadRuntime } from "./legacy-runtime/runtime-cores/local/useLocalRuntime.js";
// --- remote-thread-list ---
export { useRemoteThreadListRuntime } from "./legacy-runtime/runtime-cores/remote-thread-list/useRemoteThreadListRuntime.js";
/** @deprecated Use `useRemoteThreadListRuntime` instead. */
export { useRemoteThreadListRuntime as unstable_useRemoteThreadListRuntime } from "./legacy-runtime/runtime-cores/remote-thread-list/useRemoteThreadListRuntime.js";
export { useCloudThreadListAdapter } from "./legacy-runtime/runtime-cores/remote-thread-list/adapter/cloud.js";
/** @deprecated Use `useCloudThreadListAdapter` instead. */
export { useCloudThreadListAdapter as unstable_useCloudThreadListAdapter } from "./legacy-runtime/runtime-cores/remote-thread-list/adapter/cloud.js";
export { InMemoryThreadListAdapter } from "@assistant-ui/core";
/** @deprecated Use `InMemoryThreadListAdapter` instead. */
export { InMemoryThreadListAdapter as unstable_InMemoryThreadListAdapter } from "@assistant-ui/core";
export { ExportedMessageRepository } from "@assistant-ui/core";
// ============================================================================
// context
// ============================================================================
export * from "./context/index.js";
// ============================================================================
// model-context
// ============================================================================
// Re-export shared from core/react
export { makeAssistantTool, makeAssistantToolUI, makeAssistantDataUI, useAssistantTool, useAssistantToolUI, useAssistantDataUI, useAssistantInstructions, useInlineRender, Tools, DataRenderers, Interactables, useInteractable, makeInteractable, } from "@assistant-ui/core/react";
export { mergeModelContexts } from "@assistant-ui/core";
export { tool } from "@assistant-ui/core";
export { Suggestions } from "@assistant-ui/core/store";
// React-only (stays)
export { makeAssistantVisible } from "./model-context/makeAssistantVisible.js";
// --- model-context/registry ---
export { ModelContextRegistry } from "@assistant-ui/core";
// --- model-context/frame ---
export { AssistantFrameHost } from "@assistant-ui/core";
export { AssistantFrameProvider } from "@assistant-ui/core";
export { FRAME_MESSAGE_CHANNEL } from "@assistant-ui/core";
export { useAssistantFrameHost } from "./model-context/frame/useAssistantFrameHost.js";
// ============================================================================
// primitives
// ============================================================================
export * as ActionBarPrimitive from "./primitives/actionBar.js";
export * as ActionBarMorePrimitive from "./primitives/actionBarMore.js";
export * as AssistantModalPrimitive from "./primitives/assistantModal.js";
export * as AttachmentPrimitive from "./primitives/attachment.js";
export * as BranchPickerPrimitive from "./primitives/branchPicker.js";
export * as ChainOfThoughtPrimitive from "./primitives/chainOfThought.js";
export * as ComposerPrimitive from "./primitives/composer.js";
export * as QueueItemPrimitive from "./primitives/queueItem.js";
export * as MessagePartPrimitive from "./primitives/messagePart.js";
export * as ErrorPrimitive from "./primitives/error.js";
export * as MessagePrimitive from "./primitives/message.js";
export * as ThreadPrimitive from "./primitives/thread.js";
export * as SuggestionPrimitive from "./primitives/suggestion.js";
export * as ThreadListPrimitive from "./primitives/threadList.js";
export * as ThreadListItemPrimitive from "./primitives/threadListItem.js";
export * as ThreadListItemMorePrimitive from "./primitives/threadListItemMore.js";
export * as SelectionToolbarPrimitive from "./primitives/selectionToolbar.js";
export { useMessagePartText } from "./primitives/messagePart/useMessagePartText.js";
export { useMessagePartReasoning } from "./primitives/messagePart/useMessagePartReasoning.js";
export { useMessagePartSource } from "./primitives/messagePart/useMessagePartSource.js";
export { useMessagePartFile } from "./primitives/messagePart/useMessagePartFile.js";
export { useMessagePartImage } from "./primitives/messagePart/useMessagePartImage.js";
export { useMessagePartData } from "./primitives/messagePart/useMessagePartData.js";
export { useThreadViewportAutoScroll } from "./primitives/thread/useThreadViewportAutoScroll.js";
export { useScrollLock } from "./primitives/reasoning/useScrollLock.js";
export { useMessageQuote } from "./hooks/useMessageQuote.js";
export { useMessageTiming } from "./hooks/useMessageTiming.js";
// ============================================================================
// devtools
// ============================================================================
export { DevToolsHooks, DevToolsProviderApi } from "./devtools/DevToolsHooks.js";
// ============================================================================
// client
// ============================================================================
// Note: Tools, DataRenderers, Suggestions, SuggestionConfig already exported above from model-context
export { ModelContext as ModelContextClient } from "@assistant-ui/core/store";
export { ChainOfThoughtClient } from "@assistant-ui/core/store";
export { ExternalThread, } from "./client/ExternalThread.js";
export { InMemoryThreadList, } from "./client/InMemoryThreadList.js";
export { SingleThreadList } from "./client/SingleThreadList.js";
// ============================================================================
// internal & augmentations
// ============================================================================
export * as INTERNAL from "./internal.js";
// Unstable - mention
export { unstable_useToolMentionAdapter, } from "./unstable/useToolMentionAdapter.js";
export { useMentionContext as unstable_useMentionContext, useMentionContextOptional as unstable_useMentionContextOptional, useMentionInternalContext as unstable_useMentionInternalContext, } from "./primitives/composer/mention/index.js";
// ============================================================================
// Backwards compatibility - deprecated exports
// ============================================================================
/**
 * @deprecated Use `useAui` instead. This alias will be removed in v0.13.
 */
export { useAui as useAssistantApi } from "@assistant-ui/store";
/**
 * @deprecated Use `useAuiState` instead. This alias will be removed in v0.13.
 */
export { useAuiState as useAssistantState } from "@assistant-ui/store";
/**
 * @deprecated Use `useAuiEvent` instead. This alias will be removed in v0.13.
 */
export { useAuiEvent as useAssistantEvent } from "@assistant-ui/store";
/**
 * @deprecated Use `AuiIf` instead. This alias will be removed in v0.13.
 */
export { AuiIf as AssistantIf } from "@assistant-ui/store";
