// @assistant-ui/core/dist/react/runtimes/createMessageConverter.js
"use client";
import { useAui, useAuiState } from "@assistant-ui/store";
import { useExternalMessageConverter, convertExternalMessages, } from "./external-message-converter.js";
import { getExternalStoreMessages } from "../../runtime/utils/external-store-message.js";
export const createMessageConverter = (callback) => {
    const result = {
        useThreadMessages: ({ messages, isRunning, joinStrategy, metadata, }) => {
            return useExternalMessageConverter({
                callback,
                messages,
                isRunning,
                joinStrategy,
                metadata,
            });
        },
        toThreadMessages: (messages, isRunning = false, metadata = {}) => {
            return convertExternalMessages(messages, callback, isRunning, metadata);
        },
        toOriginalMessages: (input) => {
            const messages = getExternalStoreMessages(input);
            if (messages.length === 0)
                throw new Error("No original messages found");
            return messages;
        },
        toOriginalMessage: (input) => {
            const messages = result.toOriginalMessages(input);
            return messages[0];
        },
        useOriginalMessage: () => {
            const messageMessages = result.useOriginalMessages();
            const first = messageMessages[0];
            return first;
        },
        useOriginalMessages: () => {
            const aui = useAui();
            const partMessages = useAuiState((s) => {
                if (aui.part.source)
                    return getExternalStoreMessages(s.part);
                return undefined;
            });
            const messageMessages = useAuiState((s) => getExternalStoreMessages(s.message));
            const messages = partMessages ?? messageMessages;
            if (messages.length === 0)
                throw new Error("No original messages found");
            return messages;
        },
    };
    return result;
};
