// @assistant-ui/core/dist/react/runtimes/useLocalRuntime.js
import { useEffect, useMemo, useRef, useState } from "react";
import { AssistantRuntimeImpl, LocalRuntimeCore } from "../../internal.js";
import { useAuiState } from "@assistant-ui/store";
import { useRemoteThreadListRuntime } from "./useRemoteThreadListRuntime.js";
import { useCloudThreadListAdapter } from "./cloud/useCloudThreadListAdapter.js";
import { useRuntimeAdapters } from "./RuntimeAdapterProvider.js";
const useLocalThreadRuntime = (chatModel, { initialMessages, ...options }) => {
    const { modelContext, ...threadListAdapters } = useRuntimeAdapters() ?? {};
    const opt = {
        ...options,
        adapters: {
            ...threadListAdapters,
            ...options.adapters,
            chatModel,
        },
    };
    const [runtime] = useState(() => new LocalRuntimeCore(opt, initialMessages));
    const threadIdRef = useRef(undefined);
    threadIdRef.current = useAuiState((s) => s.threadListItem.remoteId);
    useEffect(() => {
        runtime.threads
            .getMainThreadRuntimeCore()
            .__internal_setGetThreadId(() => threadIdRef.current);
    }, [runtime]);
    useEffect(() => {
        return () => {
            runtime.threads.getMainThreadRuntimeCore().detach();
        };
    }, [runtime]);
    useEffect(() => {
        runtime.threads.getMainThreadRuntimeCore().__internal_setOptions(opt);
        runtime.threads.getMainThreadRuntimeCore().__internal_load();
    });
    useEffect(() => {
        if (!modelContext)
            return undefined;
        return runtime.registerModelContextProvider(modelContext);
    }, [modelContext, runtime]);
    return useMemo(() => new AssistantRuntimeImpl(runtime), [runtime]);
};
export const splitLocalRuntimeOptions = (options) => {
    const { cloud, initialMessages, maxSteps, adapters, unstable_humanToolNames, ...rest } = options;
    return {
        localRuntimeOptions: {
            cloud,
            initialMessages,
            maxSteps,
            adapters,
            unstable_humanToolNames,
        },
        otherOptions: rest,
    };
};
export const useLocalRuntime = (chatModel, { cloud, ...options } = {}) => {
    const cloudAdapter = useCloudThreadListAdapter({ cloud });
    return useRemoteThreadListRuntime({
        runtimeHook: function RuntimeHook() {
            return useLocalThreadRuntime(chatModel, options);
        },
        adapter: cloudAdapter,
        allowNesting: true,
    });
};
