// @assistant-ui/core/dist/react/runtimes/useRemoteThreadListRuntime.js
import { useState, useEffect, useMemo, useRef, useCallback } from "react";
import { BaseAssistantRuntimeCore } from "../../runtime/base/base-assistant-runtime-core.js";
import { AssistantRuntimeImpl } from "../../runtime/api/assistant-runtime.js";
import { RemoteThreadListThreadListRuntimeCore } from "./RemoteThreadListThreadListRuntimeCore.js";
import { useAui } from "@assistant-ui/store";
class RemoteThreadListRuntimeCore extends BaseAssistantRuntimeCore {
    threads;
    constructor(options) {
        super();
        this.threads = new RemoteThreadListThreadListRuntimeCore(options, this._contextProvider);
    }
    get RenderComponent() {
        return this.threads.__internal_RenderComponent;
    }
}
const useRemoteThreadListRuntimeImpl = (options) => {
    const [runtime] = useState(() => new RemoteThreadListRuntimeCore(options));
    useEffect(() => {
        runtime.threads.__internal_setOptions(options);
        runtime.threads.__internal_load();
    }, [runtime, options]);
    return useMemo(() => new AssistantRuntimeImpl(runtime), [runtime]);
};
export const useRemoteThreadListRuntime = (options) => {
    const runtimeHookRef = useRef(options.runtimeHook);
    runtimeHookRef.current = options.runtimeHook;
    const stableRuntimeHook = useCallback(() => {
        return runtimeHookRef.current();
    }, []);
    const stableOptions = useMemo(() => ({
        adapter: options.adapter,
        allowNesting: options.allowNesting,
        runtimeHook: stableRuntimeHook,
    }), [options.adapter, options.allowNesting, stableRuntimeHook]);
    const aui = useAui();
    const isNested = aui.threadListItem.source !== null;
    if (isNested) {
        if (!stableOptions.allowNesting) {
            throw new Error("useRemoteThreadListRuntime cannot be nested inside another RemoteThreadListRuntime. " +
                "Set allowNesting: true to allow nesting (the inner runtime will become a no-op).");
        }
        // If allowNesting is true and already inside a thread list context,
        // just call the runtimeHook directly (no-op behavior)
        return stableRuntimeHook();
    }
    return useRemoteThreadListRuntimeImpl(stableOptions);
};
