// @assistant-ui/core/dist/react/runtimes/useToolInvocations.js
import { useEffect, useRef, useState } from "react";
import { createAssistantStreamController, ToolResponse, unstable_toolResultStream, } from "assistant-stream";
import { AssistantMetaTransformStream, } from "assistant-stream/utils";
import { isJSONValueEqual } from "../../utils/json/is-json-equal.js";
const isArgsTextComplete = (argsText) => {
    try {
        JSON.parse(argsText);
        return true;
    }
    catch {
        return false;
    }
};
const parseArgsText = (argsText) => {
    try {
        return JSON.parse(argsText);
    }
    catch {
        return undefined;
    }
};
const isEquivalentCompleteArgsText = (previous, next) => {
    const previousValue = parseArgsText(previous);
    const nextValue = parseArgsText(next);
    if (previousValue === undefined || nextValue === undefined)
        return false;
    return isJSONValueEqual(previousValue, nextValue);
};
export function useToolInvocations({ state, getTools, onResult, setToolStatuses, }) {
    const lastToolStates = useRef({});
    const humanInputRef = useRef(new Map());
    const acRef = useRef(new AbortController());
    const executingCountRef = useRef(0);
    const startedExecutionToolCallIdsRef = useRef(new Set());
    const settledResolversRef = useRef([]);
    const toolCallIdAliasesRef = useRef(new Map());
    const ignoredResultToolCallIdsRef = useRef(new Set());
    const rewriteCounterRef = useRef(0);
    const getLogicalToolCallId = (toolCallId) => {
        return toolCallIdAliasesRef.current.get(toolCallId) ?? toolCallId;
    };
    const shouldIgnoreAndCleanupResult = (toolCallId) => {
        if (!ignoredResultToolCallIdsRef.current.has(toolCallId))
            return false;
        ignoredResultToolCallIdsRef.current.delete(toolCallId);
        toolCallIdAliasesRef.current.delete(toolCallId);
        return true;
    };
    const getWrappedTools = () => {
        const tools = getTools();
        if (!tools)
            return undefined;
        return Object.fromEntries(Object.entries(tools).map(([name, tool]) => {
            const execute = tool.execute;
            const streamCall = tool.streamCall;
            const wrappedTool = {
                ...tool,
                ...(execute !== undefined && {
                    execute: (...[args, context]) => execute(args, {
                        ...context,
                        toolCallId: getLogicalToolCallId(context.toolCallId),
                    }),
                }),
                ...(streamCall !== undefined && {
                    streamCall: (...[reader, context]) => streamCall(reader, {
                        ...context,
                        toolCallId: getLogicalToolCallId(context.toolCallId),
                    }),
                }),
            };
            return [name, wrappedTool];
        }));
    };
    const [controller] = useState(() => {
        const [stream, controller] = createAssistantStreamController();
        const transform = unstable_toolResultStream(getWrappedTools, () => acRef.current?.signal ?? new AbortController().signal, (toolCallId, payload) => {
            const logicalToolCallId = getLogicalToolCallId(toolCallId);
            return new Promise((resolve, reject) => {
                // Reject previous human input request if it exists
                const previous = humanInputRef.current.get(logicalToolCallId);
                if (previous) {
                    previous.reject(new Error("Human input request was superseded by a new request"));
                }
                humanInputRef.current.set(logicalToolCallId, { resolve, reject });
                setToolStatuses((prev) => ({
                    ...prev,
                    [logicalToolCallId]: {
                        type: "interrupt",
                        payload: { type: "human", payload },
                    },
                }));
            });
        }, {
            onExecutionStart: (toolCallId) => {
                if (ignoredResultToolCallIdsRef.current.has(toolCallId)) {
                    return;
                }
                startedExecutionToolCallIdsRef.current.add(toolCallId);
                const logicalToolCallId = getLogicalToolCallId(toolCallId);
                executingCountRef.current++;
                setToolStatuses((prev) => ({
                    ...prev,
                    [logicalToolCallId]: { type: "executing" },
                }));
            },
            onExecutionEnd: (toolCallId) => {
                const wasStarted = startedExecutionToolCallIdsRef.current.delete(toolCallId);
                if (ignoredResultToolCallIdsRef.current.has(toolCallId)) {
                    if (wasStarted) {
                        executingCountRef.current--;
                        if (executingCountRef.current === 0) {
                            settledResolversRef.current.forEach((resolve) => resolve());
                            settledResolversRef.current = [];
                        }
                    }
                    return;
                }
                if (!wasStarted) {
                    return;
                }
                const logicalToolCallId = getLogicalToolCallId(toolCallId);
                executingCountRef.current--;
                setToolStatuses((prev) => {
                    const next = { ...prev };
                    delete next[logicalToolCallId];
                    return next;
                });
                // Resolve any waiting abort promises when all tools have settled
                if (executingCountRef.current === 0) {
                    settledResolversRef.current.forEach((resolve) => resolve());
                    settledResolversRef.current = [];
                }
            },
        });
        stream
            .pipeThrough(transform)
            .pipeThrough(new AssistantMetaTransformStream())
            .pipeTo(new WritableStream({
            write(chunk) {
                if (chunk.type === "result") {
                    if (shouldIgnoreAndCleanupResult(chunk.meta.toolCallId)) {
                        return;
                    }
                    const logicalToolCallId = getLogicalToolCallId(chunk.meta.toolCallId);
                    if (logicalToolCallId !== chunk.meta.toolCallId) {
                        toolCallIdAliasesRef.current.delete(chunk.meta.toolCallId);
                    }
                    // the tool call result was already set by the backend
                    if (lastToolStates.current[logicalToolCallId]?.hasResult)
                        return;
                    onResult({
                        type: "add-tool-result",
                        toolCallId: logicalToolCallId,
                        toolName: chunk.meta.toolName,
                        result: chunk.result,
                        isError: chunk.isError,
                        ...(chunk.artifact && { artifact: chunk.artifact }),
                    });
                }
            },
        }));
        return controller;
    });
    const ignoredToolIds = useRef(new Set());
    const isInitialState = useRef(true);
    useEffect(() => {
        const createToolState = ({ controller, streamToolCallId, }) => ({
            argsText: "",
            hasResult: false,
            argsComplete: false,
            streamToolCallId,
            controller,
        });
        const setToolState = (toolCallId, state) => {
            lastToolStates.current[toolCallId] = state;
            return state;
        };
        const patchToolState = (toolCallId, state, patch) => {
            return setToolState(toolCallId, { ...state, ...patch });
        };
        const hasExecutableTool = (toolName) => {
            const tool = getTools()?.[toolName];
            return tool?.execute !== undefined || tool?.streamCall !== undefined;
        };
        const shouldCloseArgsStream = ({ toolName, argsText, hasResult, }) => {
            if (hasResult)
                return true;
            if (!hasExecutableTool(toolName)) {
                // Non-executable tools can emit parseable snapshots mid-stream.
                // Wait until the run settles before closing the args stream.
                return !state.isRunning && isArgsTextComplete(argsText);
            }
            return isArgsTextComplete(argsText);
        };
        const restartToolArgsStream = ({ toolCallId, toolName, state, }) => {
            ignoredResultToolCallIdsRef.current.add(state.streamToolCallId);
            state.controller.argsText.close();
            const streamToolCallId = `${toolCallId}:rewrite:${rewriteCounterRef.current++}`;
            toolCallIdAliasesRef.current.set(streamToolCallId, toolCallId);
            const toolCallController = controller.addToolCallPart({
                toolName,
                toolCallId: streamToolCallId,
            });
            if (process.env.NODE_ENV !== "production") {
                console.warn("started replacement stream tool call", {
                    toolCallId,
                    streamToolCallId,
                });
            }
            return setToolState(toolCallId, {
                ...createToolState({
                    controller: toolCallController,
                    streamToolCallId,
                }),
                hasResult: state.hasResult,
            });
        };
        const processMessages = (messages) => {
            messages.forEach((message) => {
                message.content.forEach((content) => {
                    if (content.type === "tool-call") {
                        if (isInitialState.current) {
                            ignoredToolIds.current.add(content.toolCallId);
                        }
                        else {
                            if (ignoredToolIds.current.has(content.toolCallId)) {
                                return;
                            }
                            let lastState = lastToolStates.current[content.toolCallId];
                            if (!lastState) {
                                toolCallIdAliasesRef.current.set(content.toolCallId, content.toolCallId);
                                const toolCallController = controller.addToolCallPart({
                                    toolName: content.toolName,
                                    toolCallId: content.toolCallId,
                                });
                                lastState = setToolState(content.toolCallId, createToolState({
                                    controller: toolCallController,
                                    streamToolCallId: content.toolCallId,
                                }));
                            }
                            if (content.argsText !== lastState.argsText) {
                                let shouldWriteArgsText = true;
                                if (lastState.argsComplete) {
                                    if (isEquivalentCompleteArgsText(lastState.argsText, content.argsText)) {
                                        lastState = patchToolState(content.toolCallId, lastState, {
                                            argsText: content.argsText,
                                        });
                                        shouldWriteArgsText = false;
                                    }
                                    if (shouldWriteArgsText) {
                                        const canRestartClosedArgsStream = !lastState.hasResult &&
                                            !startedExecutionToolCallIdsRef.current.has(lastState.streamToolCallId);
                                        if (process.env.NODE_ENV !== "production") {
                                            console.warn(canRestartClosedArgsStream
                                                ? "argsText updated after controller was closed, restarting tool args stream:"
                                                : "argsText updated after controller was closed:", {
                                                previous: lastState.argsText,
                                                next: content.argsText,
                                            });
                                        }
                                        if (!canRestartClosedArgsStream) {
                                            lastState = patchToolState(content.toolCallId, lastState, {
                                                argsText: content.argsText,
                                            });
                                            shouldWriteArgsText = false;
                                        }
                                    }
                                    if (shouldWriteArgsText) {
                                        lastState = restartToolArgsStream({
                                            toolCallId: content.toolCallId,
                                            toolName: content.toolName,
                                            state: lastState,
                                        });
                                    }
                                }
                                else if (!content.argsText.startsWith(lastState.argsText)) {
                                    // Check if this is key reordering (both are complete JSON)
                                    // This happens when transitioning from streaming to complete state
                                    // and the provider returns keys in a different order
                                    if (isArgsTextComplete(lastState.argsText) &&
                                        isArgsTextComplete(content.argsText) &&
                                        isEquivalentCompleteArgsText(lastState.argsText, content.argsText)) {
                                        const shouldClose = shouldCloseArgsStream({
                                            toolName: content.toolName,
                                            argsText: content.argsText,
                                            hasResult: content.result !== undefined,
                                        });
                                        if (shouldClose) {
                                            lastState.controller.argsText.close();
                                        }
                                        lastState = patchToolState(content.toolCallId, lastState, {
                                            argsText: content.argsText,
                                            argsComplete: shouldClose,
                                        });
                                        shouldWriteArgsText = false;
                                    }
                                    if (shouldWriteArgsText) {
                                        if (process.env.NODE_ENV !== "production") {
                                            console.warn("argsText rewrote previous snapshot, restarting tool args stream:", {
                                                previous: lastState.argsText,
                                                next: content.argsText,
                                                toolCallId: content.toolCallId,
                                            });
                                        }
                                        lastState = restartToolArgsStream({
                                            toolCallId: content.toolCallId,
                                            toolName: content.toolName,
                                            state: lastState,
                                        });
                                    }
                                }
                                if (shouldWriteArgsText) {
                                    const argsTextDelta = content.argsText.slice(lastState.argsText.length);
                                    lastState.controller.argsText.append(argsTextDelta);
                                    const shouldClose = shouldCloseArgsStream({
                                        toolName: content.toolName,
                                        argsText: content.argsText,
                                        hasResult: content.result !== undefined,
                                    });
                                    if (shouldClose) {
                                        lastState.controller.argsText.close();
                                    }
                                    lastState = patchToolState(content.toolCallId, lastState, {
                                        argsText: content.argsText,
                                        argsComplete: shouldClose,
                                    });
                                }
                            }
                            if (!lastState.argsComplete) {
                                const shouldClose = shouldCloseArgsStream({
                                    toolName: content.toolName,
                                    argsText: content.argsText,
                                    hasResult: content.result !== undefined,
                                });
                                if (shouldClose) {
                                    lastState.controller.argsText.close();
                                    lastState = patchToolState(content.toolCallId, lastState, {
                                        argsText: content.argsText,
                                        argsComplete: true,
                                    });
                                }
                            }
                            if (content.result !== undefined && !lastState.hasResult) {
                                patchToolState(content.toolCallId, lastState, {
                                    hasResult: true,
                                    argsComplete: true,
                                });
                                lastState.controller.setResponse(new ToolResponse({
                                    result: content.result,
                                    artifact: content.artifact,
                                    isError: content.isError,
                                }));
                                lastState.controller.close();
                            }
                        }
                        // Recursively process nested messages
                        if (content.messages) {
                            processMessages(content.messages);
                        }
                    }
                });
            });
        };
        processMessages(state.messages);
        if (isInitialState.current) {
            isInitialState.current = false;
        }
    }, [state, controller, getTools]);
    const abort = () => {
        humanInputRef.current.forEach(({ reject }) => {
            reject(new Error("Tool execution aborted"));
        });
        humanInputRef.current.clear();
        acRef.current.abort();
        acRef.current = new AbortController();
        // Return a promise that resolves when all executing tools have settled
        if (executingCountRef.current === 0) {
            return Promise.resolve();
        }
        return new Promise((resolve) => {
            settledResolversRef.current.push(resolve);
        });
    };
    return {
        reset: () => {
            isInitialState.current = true;
            void abort().finally(() => {
                startedExecutionToolCallIdsRef.current.clear();
                toolCallIdAliasesRef.current.clear();
                ignoredResultToolCallIdsRef.current.clear();
                rewriteCounterRef.current = 0;
            });
        },
        abort,
        resume: (toolCallId, payload) => {
            const handlers = humanInputRef.current.get(toolCallId);
            if (handlers) {
                humanInputRef.current.delete(toolCallId);
                setToolStatuses((prev) => ({
                    ...prev,
                    [toolCallId]: { type: "executing" },
                }));
                handlers.resolve(payload);
            }
            else {
                throw new Error(`Tool call ${toolCallId} is not waiting for human input`);
            }
        },
    };
}
