// @unkey/api/dist/esm/models/components/v2deploygetdeploymentresponsedata.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod/v3";
import { safeParse } from "../../lib/schemas.js";
import * as openEnums from "../../types/enums.js";
import { V2DeployDeploymentStep$inboundSchema, } from "./v2deploydeploymentstep.js";
/**
 * Current deployment status
 */
export const Status = {
    Unspecified: "UNSPECIFIED",
    Pending: "PENDING",
    Starting: "STARTING",
    Building: "BUILDING",
    Deploying: "DEPLOYING",
    Network: "NETWORK",
    Finalizing: "FINALIZING",
    Ready: "READY",
    Failed: "FAILED",
    Skipped: "SKIPPED",
    AwaitingApproval: "AWAITING_APPROVAL",
};
/** @internal */
export const Status$inboundSchema = openEnums.inboundSchema(Status);
/** @internal */
export const V2DeployGetDeploymentResponseData$inboundSchema = z.object({
    id: z.string(),
    status: Status$inboundSchema,
    errorMessage: z.string().optional(),
    hostnames: z.array(z.string()).optional(),
    steps: z.array(V2DeployDeploymentStep$inboundSchema).optional(),
});
export function v2DeployGetDeploymentResponseDataFromJSON(jsonString) {
    return safeParse(jsonString, (x) => V2DeployGetDeploymentResponseData$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'V2DeployGetDeploymentResponseData' from JSON`);
}
