// @unkey/api/dist/esm/models/components/v2keysverifykeyresponsedata.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod/v3";
import { safeParse } from "../../lib/schemas.js";
import * as openEnums from "../../types/enums.js";
import { Identity$inboundSchema } from "./identity.js";
import { VerifyKeyRatelimitData$inboundSchema, } from "./verifykeyratelimitdata.js";
/**
 * A machine-readable code indicating the verification status
 *
 * @remarks
 * or failure reason. Values: `VALID` (key is valid and passed all checks), `NOT_FOUND` (key doesn't
 * exist or belongs to wrong API), `FORBIDDEN` (key lacks required permissions), `INSUFFICIENT_PERMISSIONS`
 * (key lacks specific required permissions for this request), `USAGE_EXCEEDED` (key has no remaining credits), `RATE_LIMITED` (key exceeded rate limits), `DISABLED` (key was explicitly disabled),
 * `EXPIRED` (key has passed its expiration date).
 */
export const Code = {
    Valid: "VALID",
    NotFound: "NOT_FOUND",
    Forbidden: "FORBIDDEN",
    InsufficientPermissions: "INSUFFICIENT_PERMISSIONS",
    UsageExceeded: "USAGE_EXCEEDED",
    RateLimited: "RATE_LIMITED",
    Disabled: "DISABLED",
    Expired: "EXPIRED",
};
/** @internal */
export const Code$inboundSchema = openEnums.inboundSchema(Code);
/** @internal */
export const V2KeysVerifyKeyResponseData$inboundSchema = z.object({
    valid: z.boolean(),
    code: Code$inboundSchema,
    keyId: z.string().optional(),
    name: z.string().optional(),
    meta: z.record(z.any()).optional(),
    expires: z.number().int().optional(),
    credits: z.number().int().optional(),
    enabled: z.boolean().optional(),
    permissions: z.array(z.string()).optional(),
    roles: z.array(z.string()).optional(),
    identity: Identity$inboundSchema.optional(),
    ratelimits: z.array(VerifyKeyRatelimitData$inboundSchema).optional(),
});
export function v2KeysVerifyKeyResponseDataFromJSON(jsonString) {
    return safeParse(jsonString, (x) => V2KeysVerifyKeyResponseData$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'V2KeysVerifyKeyResponseData' from JSON`);
}
