// @unkey/api/dist/esm/models/components/v2keysupdatecreditsrequestbody.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod/v3";
import * as openEnums from "../../types/enums.js";
/**
 * Defines how to modify the key's remaining credits. Use 'set' to replace current credits with a specific value or unlimited usage, 'increment' to add credits for plan upgrades or credit purchases, and 'decrement' to reduce credits for refunds or policy violations.
 *
 * @remarks
 */
export const Operation = {
    Set: "set",
    Increment: "increment",
    Decrement: "decrement",
};
/** @internal */
export const Operation$outboundSchema = openEnums.outboundSchema(Operation);
/** @internal */
export const V2KeysUpdateCreditsRequestBody$outboundSchema = z.object({
    keyId: z.string(),
    value: z.nullable(z.number().int()).optional(),
    operation: Operation$outboundSchema,
});
export function v2KeysUpdateCreditsRequestBodyToJSON(v2KeysUpdateCreditsRequestBody) {
    return JSON.stringify(V2KeysUpdateCreditsRequestBody$outboundSchema.parse(v2KeysUpdateCreditsRequestBody));
}
