// @unkey/api/dist/esm/models/components/v2keyscreatekeyrequestbody.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod/v3";
import { KeyCreditsData$outboundSchema, } from "./keycreditsdata.js";
import { RatelimitRequest$outboundSchema, } from "./ratelimitrequest.js";
/** @internal */
export const V2KeysCreateKeyRequestBody$outboundSchema = z.object({
    apiId: z.string(),
    prefix: z.string().optional(),
    name: z.string().optional(),
    byteLength: z.number().int().default(16),
    externalId: z.string().optional(),
    meta: z.record(z.any()).optional(),
    roles: z.array(z.string()).optional(),
    permissions: z.array(z.string()).optional(),
    expires: z.number().int().optional(),
    credits: KeyCreditsData$outboundSchema.optional(),
    ratelimits: z.array(RatelimitRequest$outboundSchema).optional(),
    enabled: z.boolean().default(true),
    recoverable: z.boolean().default(false),
});
export function v2KeysCreateKeyRequestBodyToJSON(v2KeysCreateKeyRequestBody) {
    return JSON.stringify(V2KeysCreateKeyRequestBody$outboundSchema.parse(v2KeysCreateKeyRequestBody));
}
