// @unkey/api/dist/esm/models/components/v2keysupdatekeyrequestbody.js
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
import * as z from "zod/v3";
import { RatelimitRequest$outboundSchema, } from "./ratelimitrequest.js";
import { UpdateKeyCreditsData$outboundSchema, } from "./updatekeycreditsdata.js";
/** @internal */
export const V2KeysUpdateKeyRequestBody$outboundSchema = z.object({
    keyId: z.string(),
    name: z.nullable(z.string()).optional(),
    externalId: z.nullable(z.string()).optional(),
    meta: z.nullable(z.record(z.any())).optional(),
    expires: z.nullable(z.number().int()).optional(),
    credits: z.nullable(UpdateKeyCreditsData$outboundSchema).optional(),
    ratelimits: z.array(RatelimitRequest$outboundSchema).optional(),
    enabled: z.boolean().optional(),
    roles: z.array(z.string()).optional(),
    permissions: z.array(z.string()).optional(),
});
export function v2KeysUpdateKeyRequestBodyToJSON(v2KeysUpdateKeyRequestBody) {
    return JSON.stringify(V2KeysUpdateKeyRequestBody$outboundSchema.parse(v2KeysUpdateKeyRequestBody));
}
