// @assistant-ui/core/dist/react/primitives/message/MessageParts.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { memo, useMemo, } from "react";
import { RenderChildrenWithAccessor, useAuiState, useAui, } from "@assistant-ui/store";
import { PartByIndexProvider } from "../../providers/PartByIndexProvider.js";
import { TextMessagePartProvider } from "../../providers/TextMessagePartProvider.js";
import { ChainOfThoughtByIndicesProvider } from "../../providers/ChainOfThoughtByIndicesProvider.js";
import { getMessageQuote } from "../../utils/getMessageQuote.js";
import { useShallow } from "zustand/shallow";
/**
 * Creates a group state manager for a specific part type.
 * Returns functions to start, end, and finalize groups.
 */
const createGroupState = (groupType) => {
    let start = -1;
    return {
        startGroup: (index) => {
            if (start === -1) {
                start = index;
            }
        },
        endGroup: (endIndex, ranges) => {
            if (start !== -1) {
                ranges.push({
                    type: groupType,
                    startIndex: start,
                    endIndex,
                });
                start = -1;
            }
        },
        finalize: (endIndex, ranges) => {
            if (start !== -1) {
                ranges.push({
                    type: groupType,
                    startIndex: start,
                    endIndex,
                });
            }
        },
    };
};
/**
 * Groups consecutive tool-call and reasoning message parts into ranges.
 * Always groups tool calls and reasoning parts, even if there's only one.
 * When useChainOfThought is true, groups tool-call and reasoning parts together.
 */
const groupMessageParts = (messageTypes, useChainOfThought) => {
    const ranges = [];
    if (useChainOfThought) {
        const chainOfThoughtGroup = createGroupState("chainOfThoughtGroup");
        for (let i = 0; i < messageTypes.length; i++) {
            const type = messageTypes[i];
            if (type === "tool-call" || type === "reasoning") {
                chainOfThoughtGroup.startGroup(i);
            }
            else {
                chainOfThoughtGroup.endGroup(i - 1, ranges);
                ranges.push({ type: "single", index: i });
            }
        }
        chainOfThoughtGroup.finalize(messageTypes.length - 1, ranges);
    }
    else {
        const toolGroup = createGroupState("toolGroup");
        const reasoningGroup = createGroupState("reasoningGroup");
        for (let i = 0; i < messageTypes.length; i++) {
            const type = messageTypes[i];
            if (type === "tool-call") {
                reasoningGroup.endGroup(i - 1, ranges);
                toolGroup.startGroup(i);
            }
            else if (type === "reasoning") {
                toolGroup.endGroup(i - 1, ranges);
                reasoningGroup.startGroup(i);
            }
            else {
                toolGroup.endGroup(i - 1, ranges);
                reasoningGroup.endGroup(i - 1, ranges);
                ranges.push({ type: "single", index: i });
            }
        }
        toolGroup.finalize(messageTypes.length - 1, ranges);
        reasoningGroup.finalize(messageTypes.length - 1, ranges);
    }
    return ranges;
};
const useMessagePartsGroups = (useChainOfThought) => {
    const messageTypes = useAuiState(useShallow((s) => s.message.parts.map((c) => c.type)));
    return useMemo(() => {
        if (messageTypes.length === 0) {
            return [];
        }
        return groupMessageParts(messageTypes, useChainOfThought);
    }, [messageTypes, useChainOfThought]);
};
const ToolUIDisplay = ({ Fallback, ...props }) => {
    const Render = useAuiState((s) => {
        const Render = s.tools.tools[props.toolName] ?? Fallback;
        if (Array.isArray(Render))
            return Render[0] ?? Fallback;
        return Render;
    });
    if (!Render)
        return null;
    return _jsx(Render, { ...props });
};
const DataUIDisplay = ({ Fallback, ...props }) => {
    const Render = useAuiState((s) => {
        const Render = s.dataRenderers.renderers[props.name] ?? Fallback;
        if (Array.isArray(Render))
            return Render[0] ?? Fallback;
        return Render;
    });
    if (!Render)
        return null;
    return _jsx(Render, { ...props });
};
/**
 * Platform-agnostic no-op default components.
 * Each platform (web, RN) wraps MessagePrimitiveParts with its own defaults.
 */
export const defaultComponents = {
    Text: () => null,
    Reasoning: () => null,
    Source: () => null,
    Image: () => null,
    File: () => null,
    Unstable_Audio: () => null,
    ToolGroup: ({ children }) => children,
    ReasoningGroup: ({ children }) => children,
};
export const MessagePartComponent = ({ components: { Text = defaultComponents.Text, Reasoning = defaultComponents.Reasoning, Image = defaultComponents.Image, Source = defaultComponents.Source, File = defaultComponents.File, Unstable_Audio: Audio = defaultComponents.Unstable_Audio, tools = {}, data, } = {}, }) => {
    const aui = useAui();
    const part = useAuiState((s) => s.part);
    const type = part.type;
    if (type === "tool-call") {
        const addResult = aui.part().addToolResult;
        const resume = aui.part().resumeToolCall;
        if ("Override" in tools)
            return _jsx(tools.Override, { ...part, addResult: addResult, resume: resume });
        const Tool = tools.by_name?.[part.toolName] ?? tools.Fallback;
        return (_jsx(ToolUIDisplay, { ...part, Fallback: Tool, addResult: addResult, resume: resume }));
    }
    if (part.status?.type === "requires-action")
        throw new Error("Encountered unexpected requires-action status");
    switch (type) {
        case "text":
            return _jsx(Text, { ...part });
        case "reasoning":
            return _jsx(Reasoning, { ...part });
        case "source":
            return _jsx(Source, { ...part });
        case "image":
            return _jsx(Image, { ...part });
        case "file":
            return _jsx(File, { ...part });
        case "audio":
            return _jsx(Audio, { ...part });
        case "data": {
            const Data = data?.by_name?.[part.name] ?? data?.Fallback;
            return _jsx(DataUIDisplay, { ...part, Fallback: Data });
        }
        default:
            console.warn(`Unknown message part type: ${type}`);
            return null;
    }
};
/**
 * Renders a single message part at the specified index.
 */
export const MessagePrimitivePartByIndex = memo(({ index, components }) => {
    return (_jsx(PartByIndexProvider, { index: index, children: _jsx(MessagePartComponent, { components: components }) }));
}, (prev, next) => prev.index === next.index &&
    prev.components?.Text === next.components?.Text &&
    prev.components?.Reasoning === next.components?.Reasoning &&
    prev.components?.Source === next.components?.Source &&
    prev.components?.Image === next.components?.Image &&
    prev.components?.File === next.components?.File &&
    prev.components?.Unstable_Audio === next.components?.Unstable_Audio &&
    prev.components?.tools === next.components?.tools &&
    prev.components?.data === next.components?.data &&
    prev.components?.ToolGroup === next.components?.ToolGroup &&
    prev.components?.ReasoningGroup === next.components?.ReasoningGroup);
MessagePrimitivePartByIndex.displayName = "MessagePrimitive.PartByIndex";
const EmptyPartFallback = ({ status, component: Component }) => {
    return (_jsx(TextMessagePartProvider, { text: "", isRunning: status.type === "running", children: _jsx(Component, { type: "text", text: "", status: status }) }));
};
const COMPLETE_STATUS = Object.freeze({
    type: "complete",
});
const EmptyPartsImpl = ({ components }) => {
    const status = useAuiState((s) => (s.message.status ?? COMPLETE_STATUS));
    if (components?.Empty)
        return _jsx(components.Empty, { status: status });
    if (status.type !== "running")
        return null;
    return (_jsx(EmptyPartFallback, { status: status, component: components?.Text ?? defaultComponents.Text }));
};
const EmptyParts = memo(EmptyPartsImpl, (prev, next) => prev.components?.Empty === next.components?.Empty &&
    prev.components?.Text === next.components?.Text);
const ConditionalEmptyImpl = ({ components, enabled }) => {
    const shouldShowEmpty = useAuiState((s) => {
        if (!enabled)
            return false;
        if (s.message.parts.length === 0)
            return false;
        const lastPart = s.message.parts[s.message.parts.length - 1];
        return lastPart?.type !== "text" && lastPart?.type !== "reasoning";
    });
    if (!shouldShowEmpty)
        return null;
    return _jsx(EmptyParts, { components: components });
};
const ConditionalEmpty = memo(ConditionalEmptyImpl, (prev, next) => prev.enabled === next.enabled &&
    prev.components?.Empty === next.components?.Empty &&
    prev.components?.Text === next.components?.Text);
const QuoteRendererImpl = ({ Quote, }) => {
    const quoteInfo = useAuiState(getMessageQuote);
    if (!quoteInfo)
        return null;
    return _jsx(Quote, { text: quoteInfo.text, messageId: quoteInfo.messageId });
};
const QuoteRenderer = memo(QuoteRendererImpl);
/**
 * Stable propless component that renders the registered tool UI for the
 * current part context. Reads tool registry and part state from context.
 */
const RegisteredToolUI = () => {
    const aui = useAui();
    const part = useAuiState((s) => s.part);
    const Render = useAuiState((s) => {
        if (s.part.type !== "tool-call")
            return null;
        const entry = s.tools.tools[s.part.toolName];
        if (Array.isArray(entry))
            return entry[0] ?? null;
        return entry ?? null;
    });
    if (!Render || part.type !== "tool-call")
        return null;
    return (_jsx(Render, { ...part, addResult: aui.part().addToolResult, resume: aui.part().resumeToolCall }));
};
/**
 * Stable propless component that renders the registered data renderer UI
 * for the current part context.
 */
const RegisteredDataRendererUI = () => {
    const part = useAuiState((s) => s.part);
    const Render = useAuiState((s) => {
        if (s.part.type !== "data")
            return null;
        const entry = s.dataRenderers.renderers[s.part.name];
        if (Array.isArray(entry))
            return entry[0] ?? null;
        return entry ?? null;
    });
    if (!Render || part.type !== "data")
        return null;
    return _jsx(Render, { ...part });
};
/**
 * Fallback component rendered when the children render function returns null.
 * Renders registered tool/data UIs via context.
 * For all other part types, renders nothing.
 *
 * This allows users to write:
 *   {({ part }) => {
 *     if (part.type === "text") return <MyText />;
 *     return null; // tool UIs and data UIs still render via registry
 *   }}
 *
 * To explicitly render nothing (suppressing registered UIs), return <></>.
 */
const DefaultPartFallback = () => {
    const partType = useAuiState((s) => s.part.type);
    if (partType === "tool-call")
        return _jsx(RegisteredToolUI, {});
    if (partType === "data")
        return _jsx(RegisteredDataRendererUI, {});
    return null;
};
const MessagePrimitivePartsInner = ({ children }) => {
    const aui = useAui();
    const contentLength = useAuiState((s) => s.message.parts.length);
    // biome-ignore lint/correctness/useExhaustiveDependencies: aui accessors are stable refs
    return useMemo(() => Array.from({ length: contentLength }, (_, index) => (_jsx(PartByIndexProvider, { index: index, children: _jsx(RenderChildrenWithAccessor, { getItemState: (aui) => aui.message().part({ index }).getState(), children: (getItem) => {
                const result = children({
                    get part() {
                        const state = getItem();
                        if (state.type === "tool-call") {
                            const entry = aui.tools().getState().tools[state.toolName];
                            const hasUI = Array.isArray(entry) ? !!entry[0] : !!entry;
                            const partMethods = aui.message().part({ index });
                            return {
                                ...state,
                                toolUI: hasUI ? _jsx(RegisteredToolUI, {}) : null,
                                addResult: partMethods.addToolResult,
                                resume: partMethods.resumeToolCall,
                            };
                        }
                        if (state.type === "data") {
                            const entry = aui.dataRenderers().getState().renderers[state.name];
                            const hasUI = Array.isArray(entry) ? !!entry[0] : !!entry;
                            return {
                                ...state,
                                dataRendererUI: hasUI ? (_jsx(RegisteredDataRendererUI, {})) : null,
                            };
                        }
                        return state;
                    },
                });
                if (result !== null)
                    return result;
                return _jsx(DefaultPartFallback, {});
            } }) }, index))), [contentLength, children]);
};
/**
 * Renders the parts of a message with support for multiple content types.
 *
 * This is the platform-agnostic base. Each platform wraps this with its own
 * default components (web uses `<p>`, `<span>`; RN would use `<Text>`, etc.).
 */
export const MessagePrimitiveParts = ({ components, unstable_showEmptyOnNonTextEnd = true, children, }) => {
    if (children) {
        return _jsx(MessagePrimitivePartsInner, { children: children });
    }
    return (_jsx(MessagePrimitivePartsCompat, { components: components, unstable_showEmptyOnNonTextEnd: unstable_showEmptyOnNonTextEnd }));
};
MessagePrimitiveParts.displayName = "MessagePrimitive.Parts";
const MessagePrimitivePartsCompat = ({ components, unstable_showEmptyOnNonTextEnd }) => {
    const contentLength = useAuiState((s) => s.message.parts.length);
    const useChainOfThought = !!components?.ChainOfThought;
    const messageRanges = useMessagePartsGroups(useChainOfThought);
    const partsElements = useMemo(() => {
        if (contentLength === 0) {
            return _jsx(EmptyParts, { components: components });
        }
        return messageRanges.map((range) => {
            if (range.type === "single") {
                return (_jsx(MessagePrimitivePartByIndex, { index: range.index, components: components }, range.index));
            }
            else if (range.type === "chainOfThoughtGroup") {
                const ChainOfThoughtComponent = components?.ChainOfThought;
                if (!ChainOfThoughtComponent)
                    return null;
                return (_jsx(ChainOfThoughtByIndicesProvider, { startIndex: range.startIndex, endIndex: range.endIndex, children: _jsx(ChainOfThoughtComponent, {}) }, `chainOfThought-${range.startIndex}`));
            }
            else if (range.type === "toolGroup") {
                const ToolGroupComponent = components?.ToolGroup ?? defaultComponents.ToolGroup;
                return (_jsx(ToolGroupComponent, { startIndex: range.startIndex, endIndex: range.endIndex, children: Array.from({ length: range.endIndex - range.startIndex + 1 }, (_, i) => (_jsx(MessagePrimitivePartByIndex, { index: range.startIndex + i, components: components }, i))) }, `tool-${range.startIndex}`));
            }
            else {
                // reasoningGroup
                const ReasoningGroupComponent = components?.ReasoningGroup ?? defaultComponents.ReasoningGroup;
                return (_jsx(ReasoningGroupComponent, { startIndex: range.startIndex, endIndex: range.endIndex, children: Array.from({ length: range.endIndex - range.startIndex + 1 }, (_, i) => (_jsx(MessagePrimitivePartByIndex, { index: range.startIndex + i, components: components }, i))) }, `reasoning-${range.startIndex}`));
            }
        });
    }, [messageRanges, components, contentLength]);
    return (_jsxs(_Fragment, { children: [components?.Quote && _jsx(QuoteRenderer, { Quote: components.Quote }), partsElements, _jsx(ConditionalEmpty, { components: components, enabled: unstable_showEmptyOnNonTextEnd })] }));
};
