// @unkey/api/dist/commonjs/sdk/sdk.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Unkey = void 0;
const sdks_js_1 = require("../lib/sdks.js");
const analytics_js_1 = require("./analytics.js");
const apis_js_1 = require("./apis.js");
const identities_js_1 = require("./identities.js");
const internal_js_1 = require("./internal.js");
const keys_js_1 = require("./keys.js");
const permissions_js_1 = require("./permissions.js");
const ratelimit_js_1 = require("./ratelimit.js");
class Unkey extends sdks_js_1.ClientSDK {
    get analytics() {
        return (this._analytics ?? (this._analytics = new analytics_js_1.Analytics(this._options)));
    }
    get apis() {
        return (this._apis ?? (this._apis = new apis_js_1.Apis(this._options)));
    }
    get internal() {
        return (this._internal ?? (this._internal = new internal_js_1.Internal(this._options)));
    }
    get identities() {
        return (this._identities ?? (this._identities = new identities_js_1.Identities(this._options)));
    }
    get keys() {
        return (this._keys ?? (this._keys = new keys_js_1.Keys(this._options)));
    }
    get permissions() {
        return (this._permissions ?? (this._permissions = new permissions_js_1.Permissions(this._options)));
    }
    get ratelimit() {
        return (this._ratelimit ?? (this._ratelimit = new ratelimit_js_1.Ratelimit(this._options)));
    }
}
exports.Unkey = Unkey;
