// @unkey/api/dist/commonjs/models/components/keycreditsrefill.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyCreditsRefill$outboundSchema = exports.KeyCreditsRefill$inboundSchema = exports.KeyCreditsRefillInterval$outboundSchema = exports.KeyCreditsRefillInterval$inboundSchema = exports.KeyCreditsRefillInterval = void 0;
exports.keyCreditsRefillToJSON = keyCreditsRefillToJSON;
exports.keyCreditsRefillFromJSON = keyCreditsRefillFromJSON;
const z = __importStar(require("zod/v3"));
const schemas_js_1 = require("../../lib/schemas.js");
const openEnums = __importStar(require("../../types/enums.js"));
/**
 * How often credits are automatically refilled.
 */
exports.KeyCreditsRefillInterval = {
    Daily: "daily",
    Monthly: "monthly",
};
/** @internal */
exports.KeyCreditsRefillInterval$inboundSchema = openEnums.inboundSchema(exports.KeyCreditsRefillInterval);
/** @internal */
exports.KeyCreditsRefillInterval$outboundSchema = openEnums.outboundSchema(exports.KeyCreditsRefillInterval);
/** @internal */
exports.KeyCreditsRefill$inboundSchema = z.object({
    interval: exports.KeyCreditsRefillInterval$inboundSchema,
    amount: z.number().int(),
    refillDay: z.number().int().optional(),
});
/** @internal */
exports.KeyCreditsRefill$outboundSchema = z.object({
    interval: exports.KeyCreditsRefillInterval$outboundSchema,
    amount: z.number().int(),
    refillDay: z.number().int().optional(),
});
function keyCreditsRefillToJSON(keyCreditsRefill) {
    return JSON.stringify(exports.KeyCreditsRefill$outboundSchema.parse(keyCreditsRefill));
}
function keyCreditsRefillFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.KeyCreditsRefill$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'KeyCreditsRefill' from JSON`);
}
