// @unkey/api/dist/commonjs/models/components/keyresponsedata.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyResponseData$inboundSchema = void 0;
exports.keyResponseDataFromJSON = keyResponseDataFromJSON;
const z = __importStar(require("zod/v3"));
const schemas_js_1 = require("../../lib/schemas.js");
const identity_js_1 = require("./identity.js");
const keycreditsdata_js_1 = require("./keycreditsdata.js");
const ratelimitresponse_js_1 = require("./ratelimitresponse.js");
/** @internal */
exports.KeyResponseData$inboundSchema = z.object({
    keyId: z.string(),
    start: z.string(),
    enabled: z.boolean(),
    name: z.string().optional(),
    meta: z.record(z.any()).optional(),
    createdAt: z.number().int(),
    updatedAt: z.number().int().optional(),
    lastUsedAt: z.number().int().optional(),
    expires: z.number().int().optional(),
    permissions: z.array(z.string()).optional(),
    roles: z.array(z.string()).optional(),
    credits: keycreditsdata_js_1.KeyCreditsData$inboundSchema.optional(),
    identity: identity_js_1.Identity$inboundSchema.optional(),
    plaintext: z.string().optional(),
    ratelimits: z.array(ratelimitresponse_js_1.RatelimitResponse$inboundSchema).optional(),
});
function keyResponseDataFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.KeyResponseData$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'KeyResponseData' from JSON`);
}
