// @unkey/api/dist/commonjs/models/components/v2deploygetdeploymentresponsedata.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.V2DeployGetDeploymentResponseData$inboundSchema = exports.Status$inboundSchema = exports.Status = void 0;
exports.v2DeployGetDeploymentResponseDataFromJSON = v2DeployGetDeploymentResponseDataFromJSON;
const z = __importStar(require("zod/v3"));
const schemas_js_1 = require("../../lib/schemas.js");
const openEnums = __importStar(require("../../types/enums.js"));
const v2deploydeploymentstep_js_1 = require("./v2deploydeploymentstep.js");
/**
 * Current deployment status
 */
exports.Status = {
    Unspecified: "UNSPECIFIED",
    Pending: "PENDING",
    Starting: "STARTING",
    Building: "BUILDING",
    Deploying: "DEPLOYING",
    Network: "NETWORK",
    Finalizing: "FINALIZING",
    Ready: "READY",
    Failed: "FAILED",
    Skipped: "SKIPPED",
    AwaitingApproval: "AWAITING_APPROVAL",
};
/** @internal */
exports.Status$inboundSchema = openEnums.inboundSchema(exports.Status);
/** @internal */
exports.V2DeployGetDeploymentResponseData$inboundSchema = z.object({
    id: z.string(),
    status: exports.Status$inboundSchema,
    errorMessage: z.string().optional(),
    hostnames: z.array(z.string()).optional(),
    steps: z.array(v2deploydeploymentstep_js_1.V2DeployDeploymentStep$inboundSchema).optional(),
});
function v2DeployGetDeploymentResponseDataFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.V2DeployGetDeploymentResponseData$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'V2DeployGetDeploymentResponseData' from JSON`);
}
