// @unkey/api/dist/commonjs/models/components/v2keyscreatekeyrequestbody.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.V2KeysCreateKeyRequestBody$outboundSchema = void 0;
exports.v2KeysCreateKeyRequestBodyToJSON = v2KeysCreateKeyRequestBodyToJSON;
const z = __importStar(require("zod/v3"));
const keycreditsdata_js_1 = require("./keycreditsdata.js");
const ratelimitrequest_js_1 = require("./ratelimitrequest.js");
/** @internal */
exports.V2KeysCreateKeyRequestBody$outboundSchema = z.object({
    apiId: z.string(),
    prefix: z.string().optional(),
    name: z.string().optional(),
    byteLength: z.number().int().default(16),
    externalId: z.string().optional(),
    meta: z.record(z.any()).optional(),
    roles: z.array(z.string()).optional(),
    permissions: z.array(z.string()).optional(),
    expires: z.number().int().optional(),
    credits: keycreditsdata_js_1.KeyCreditsData$outboundSchema.optional(),
    ratelimits: z.array(ratelimitrequest_js_1.RatelimitRequest$outboundSchema).optional(),
    enabled: z.boolean().default(true),
    recoverable: z.boolean().default(false),
});
function v2KeysCreateKeyRequestBodyToJSON(v2KeysCreateKeyRequestBody) {
    return JSON.stringify(exports.V2KeysCreateKeyRequestBody$outboundSchema.parse(v2KeysCreateKeyRequestBody));
}
