// @unkey/api/dist/commonjs/models/components/v2keysverifykeyrequestbody.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.V2KeysVerifyKeyRequestBody$outboundSchema = void 0;
exports.v2KeysVerifyKeyRequestBodyToJSON = v2KeysVerifyKeyRequestBodyToJSON;
const z = __importStar(require("zod/v3"));
const keysverifykeycredits_js_1 = require("./keysverifykeycredits.js");
const keysverifykeyratelimit_js_1 = require("./keysverifykeyratelimit.js");
/** @internal */
exports.V2KeysVerifyKeyRequestBody$outboundSchema = z.object({
    key: z.string(),
    tags: z.array(z.string()).optional(),
    permissions: z.string().optional(),
    credits: keysverifykeycredits_js_1.KeysVerifyKeyCredits$outboundSchema.optional(),
    ratelimits: z.array(keysverifykeyratelimit_js_1.KeysVerifyKeyRatelimit$outboundSchema).optional(),
    migrationId: z.string().optional(),
});
function v2KeysVerifyKeyRequestBodyToJSON(v2KeysVerifyKeyRequestBody) {
    return JSON.stringify(exports.V2KeysVerifyKeyRequestBody$outboundSchema.parse(v2KeysVerifyKeyRequestBody));
}
